{
 "cells": [
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# Mutation rates analysis\n",
    "\n",
    "In this notebook, we calculate mutation rates in PAN027 assembly in regard to GIAB stratification of easy (high-confidence) and difficult regions,\n",
    "as well as problematic regions in PAN027."
   ],
   "id": "373e007c3f94b4b7"
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "88ffb6d2-b8b2-48b4-bc22-a18f766608b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import tempfile\n",
    "import subprocess\n",
    "from pathlib import Path\n",
    "from concurrent.futures import ProcessPoolExecutor, as_completed\n",
    "import shutil"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "29d9d144-261a-4d93-b52c-3717fb464f33",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ---------- INPUTS ----------\n",
    "variants_vcf: str = \"variants.vcf.gz\"           # concatenated variant files from the second step of the switch_errors pipeline,\n",
    "                                                #  filtered not to include variants coming from problematic regions\n",
    "sw_prone_bed: str = \"sw_prone_regions.bed\"      # switch error-prone regions detected by switch_errors pipeline\n",
    "flagger_ont: str = \"flagger.PAN027.ONT.bed\"     # problematic regions on PAN027 assembly\n",
    "flagger_hifi: str = \"flagger.PAN027.hifi.bed\"\n",
    "problematic: str = \"problematic.PAN027.bed\"\n",
    "ignore_snvs_regions: list[str] = []             # ignored BED regions, not used\n",
    "UNRELIABLE_GRANDPARENTS: str = \"problematic_grandparents.vcf.gz\"    # variants originating from problematic grandparents' regions\n",
    "mat_easy: str = \"chm13_to_PAN027_mat.easy.mapped.PAN027names.bed\"   # easy regions on maternal haplotype of PAN027, creating by lifting-over\n",
    "                                                                    #  CHM13_notinalldifficultregions.bed.gz to PAN027 maternal\n",
    "                                                                    #  https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/release/genome-stratifications/v3.6/CHM13@all/Union/CHM13_notinalldifficultregions.bed.gz\n",
    "mat_hard: str = \"chm13_to_PAN027_mat.hard.PAN027names.bed\"          # difficult regions on maternal haplotype of PAN027, created by complementing\n",
    "                                                                    #  easy regions\n",
    "pat_easy: str = \"chm13_to_PAN027_pat.easy.mapped.PAN027names.bed\"   # easy regions on paternal haplotype of PAN027, created as above\n",
    "pat_hard: str = \"chm13_to_PAN027_pat.hard.PAN027names.bed\"          # difficult regions on paternal haplotype of PAN027, created as above\n",
    "all_blocks: str = \"all_blocks.bed\"  # regions covered by transmitted blocks, extracted from results.json of swirtch_errors pipeline\n",
    "\n",
    "chromosomes: list[str] = [f\"chr{i}\" for i in range(1, 23)] + [\"chrX\"]\n",
    "PAD:int = 20_000  # padding (slope) for problematic regions in PAN027\n",
    "THREADS: int = 8  # parallelism"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "db828540-43d5-4f23-afcc-e5f4e0da0236",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run(cmd: list[str], capture: bool = False, check: bool = True, shell: bool = True) -> str | None:\n",
    "    \"\"\"\n",
    "    Safely run shell command\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    cmd        - command to be executed\n",
    "    capture    - if True, capture and return stdout and stderr\n",
    "    check      - if True, verify command has executed successfully (exit code 0), otherwise throw exception\n",
    "    shell      - if True, execute in local shell\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    String stdout + stderr if capture is True, None otherwise\n",
    "    \"\"\"\n",
    "\n",
    "    if capture:\n",
    "        out = subprocess.check_output(cmd, shell=shell, stderr=subprocess.STDOUT)\n",
    "        return out.decode().strip()\n",
    "    else:\n",
    "        subprocess.run(cmd, shell=shell, check=check)\n",
    "\n",
    "def ensure_file_exists(path: str) -> None:\n",
    "    \"\"\"\n",
    "    Verifies that file exists, throws otherwise\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    path - to be verified\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    None\n",
    "    \"\"\"\n",
    "    if not Path(path).exists():\n",
    "        raise FileNotFoundError(f\"Required file not found: {path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "69075a3d-348b-4d61-8027-0fa0ef5c8c7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_inputs_for_chrom(chrom: str, tmpdir: Path) -> tuple[Path, Path]:\n",
    "    \"\"\"\n",
    "    Filter variants.vcf.gz and all_blocks.bed for given chromosome, also with removing of unreliable variants\n",
    "    Returns per-chromosome variants file and all_blocks file\n",
    "    \"\"\"\n",
    "\n",
    "    if chrom is None:\n",
    "        return variants_vcf, all_blocks  # use originals\n",
    "\n",
    "    chr_variants: Path = tmpdir / f\"variants_{chrom}.vcf.gz\"\n",
    "    chr_blocks: Path = tmpdir / f\"blocks_{chrom}.bed\"\n",
    "\n",
    "    # remove unreliable variants\n",
    "    filtered_vcf: Path = tmpdir / f\"variants_filtered_unreliable.vcf.gz\"\n",
    "    cmd_filter_unreliable = (\n",
    "        f\"bcftools isec -C -O z -o {filtered_vcf} \"\n",
    "        f\"{variants_vcf} {UNRELIABLE_GRANDPARENTS}\" # {UNRELIABLE_CONFIDENCE}\n",
    "    )\n",
    "    subprocess.run(cmd_filter_unreliable, shell=True, check=True)\n",
    "    subprocess.run(f\"bcftools index -f {filtered_vcf}\", shell=True, check=True)\n",
    "\n",
    "    # filter by chromosome\n",
    "    cmd_vcf = (\n",
    "        f\"bcftools view -r PAN027.{chrom}.maternal,PAN027.{chrom}.paternal \"\n",
    "        f\"-Oz -o {chr_variants} {filtered_vcf}\"\n",
    "    )\n",
    "    subprocess.run(cmd_vcf, shell=True, check=True)\n",
    "    subprocess.run(f\"bcftools index -f {chr_variants}\", shell=True, check=True)\n",
    "\n",
    "    # filter BED\n",
    "    cmd_bed = (\n",
    "        f\"grep -E 'PAN027\\\\.{chrom}\\\\.(maternal|paternal)' {all_blocks} > {chr_blocks}\"\n",
    "    )\n",
    "    subprocess.run(cmd_bed, shell=True, check=True)\n",
    "\n",
    "    return chr_variants, chr_blocks\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ab344dc9-f8a6-415b-993b-9bf38737b765",
   "metadata": {},
   "outputs": [],
   "source": [
    "def mk_summary_single(tmpdir: Path, variants: Path, allblocks: Path) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Summarize mutation rates per easy/difficult regions for specific input files\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    tmpdir      - temporary directory for intermediates\n",
    "    variants    - variant file (possibly per-chromosome)\n",
    "    allblocks   - transmitted blocks file (possibly per-chromosome)\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    dataframe containing mutation rates\n",
    "    \"\"\"\n",
    "\n",
    "    out_prefix: str = \"mutation_rates_summary\"\n",
    "    out_tsv: str = f\"{tmpdir}/{out_prefix}.tsv\"\n",
    "    \n",
    "    try:\n",
    "        # ---------- Build combined exclusion BED ----------\n",
    "        # 1) pad flagger_ont, flagger_hifi, and problematic\n",
    "        padded_flagger_ont: Path = tmpdir/\"flagger_ont.padded.bed\"\n",
    "        padded_flagger_hifi: Path = tmpdir/\"flagger_hifi.padded.bed\"\n",
    "        padded_problematic: Path = tmpdir/\"problematic.padded.bed\"\n",
    "\n",
    "        for inp, out in [(flagger_ont, padded_flagger_ont),\n",
    "                         (flagger_hifi, padded_flagger_hifi),\n",
    "                         (problematic, padded_problematic)]:\n",
    "            cmd = (\n",
    "                f\"awk 'BEGIN{{OFS=\\\"\\\\t\\\"}} \"\n",
    "                f\"{{s=$2-{PAD}; if(s<0) s=0; print $1, s, $3+{PAD}}}' {inp} | \"\n",
    "                f\"sort -k1,1 -k2,2n | bedtools merge -i - > {out}\"\n",
    "            )\n",
    "            print(\"[CMD] padding+merge ->\", out)\n",
    "            run(cmd)\n",
    "        \n",
    "        # 2) combine sw_prone + padded flaggers + padded problematic\n",
    "        combined_exclude: Path = tmpdir / \"combined_exclude.bed\"\n",
    "        tmp_comb: Path = tmpdir / \"combined_exclude.concat.bed\"\n",
    "        with open(tmp_comb, \"w\") as out_f:\n",
    "            for src in [sw_prone_bed, padded_flagger_ont, padded_flagger_hifi, padded_problematic]:\n",
    "                with open(src) as f:\n",
    "                    for line in f:\n",
    "                        if line.startswith(\"#\") or not line.strip():\n",
    "                            continue\n",
    "                        fields: list[str] = line.strip().split()\n",
    "                        if len(fields) < 3:\n",
    "                            continue\n",
    "                        try:\n",
    "                            start: int = int(fields[1])\n",
    "                            end: int = int(fields[2])\n",
    "                        except ValueError:\n",
    "                            continue  # skip malformed\n",
    "                        out_f.write(f\"{fields[0]}\\t{start}\\t{end}\\n\")\n",
    "        cmd = (\n",
    "            f\"sort -k1,1 -k2,2n {tmp_comb} > {tmp_comb}.sorted && \"\n",
    "            f\"bedtools merge -i {tmp_comb}.sorted > {combined_exclude}\"\n",
    "        )\n",
    "        print(\"[CMD] combining excludes ->\", combined_exclude)\n",
    "        run(cmd)\n",
    "    \n",
    "        # ---------- Prepare region sets: all / easy / hard, for maternal/paternal/all ----------\n",
    "        # output: regions_all.bed, regions_mat_easy.bed, regions_mat_hard.bed, regions_pat_easy.bed, regions_pat_hard.bed\n",
    "        # we are working only on transmitted blocks w.r.t. exclusion created above\n",
    "        def prepare_region(in_bed: str, out_bed: str) -> None: # intersect with all_blocks, subtract excludes, sort+merge\n",
    "            cmd = (\n",
    "                f\"bedtools intersect -a {in_bed} -b {allblocks} -u | \"\n",
    "                f\"bedtools subtract -a - -b {combined_exclude} | \"\n",
    "                f\"sort -k1,1 -k2,2n | bedtools merge -i - > {out_bed}\"\n",
    "            )\n",
    "            run(cmd)\n",
    "\n",
    "        def get_bed_total_length(bed_path: str) -> int:\n",
    "            \"\"\"\n",
    "            Calculates total length of regions covered by BED file\n",
    "            \"\"\"\n",
    "            cmd = f\"awk '{{s+=$3-$2}} END{{print s+0}}' {bed_path}\"\n",
    "            out = run(cmd, capture=True)\n",
    "            return int(float(out))\n",
    "    \n",
    "        regions = {}\n",
    "        # maternal easy/hard\n",
    "        regions['mat_easy'] = tmpdir/\"mat_easy.selected.bed\"\n",
    "        regions['mat_hard'] = tmpdir/\"mat_hard.selected.bed\"\n",
    "        prepare_region(mat_easy, regions['mat_easy'])\n",
    "        prepare_region(mat_hard, regions['mat_hard'])\n",
    "        # paternal easy/hard\n",
    "        regions['pat_easy'] = tmpdir/\"pat_easy.selected.bed\"\n",
    "        regions['pat_hard'] = tmpdir/\"pat_hard.selected.bed\"\n",
    "        prepare_region(pat_easy, regions['pat_easy'])\n",
    "        prepare_region(pat_hard, regions['pat_hard'])\n",
    "    \n",
    "        # define maternal-only = mat intersect all_blocks, paternal-only similarly\n",
    "        mat_union: Path = tmpdir/\"mat_union.bed\"\n",
    "        pat_union: Path = tmpdir/\"pat_union.bed\"\n",
    "        run(f\"cat {regions['mat_easy']} {regions['mat_hard']} | sort -k1,1 -k2,2n | bedtools merge -i - > {mat_union}\")\n",
    "        run(f\"cat {regions['pat_easy']} {regions['pat_hard']} | sort -k1,1 -k2,2n | bedtools merge -i - > {pat_union}\")\n",
    "    \n",
    "        mat_only: Path = tmpdir/\"mat_only.bed\"\n",
    "        pat_only: Path = tmpdir/\"pat_only.bed\"\n",
    "        run(f\"bedtools subtract -a {mat_union} -b {pat_union} | sort -k1,1 -k2,2n | bedtools merge -i - > {mat_only}\")\n",
    "        run(f\"bedtools subtract -a {pat_union} -b {mat_union} | sort -k1,1 -k2,2n | bedtools merge -i - > {pat_only}\")\n",
    "    \n",
    "        all_regions: Path = tmpdir/\"all_regions.bed\"\n",
    "        run(f\"cat {mat_union} {pat_union} | sort -k1,1 -k2,2n | bedtools merge -i - > {all_regions}\")\n",
    "\n",
    "        mat_len_total: int = get_bed_total_length(mat_only)\n",
    "        pat_len_total: int = get_bed_total_length(pat_only)\n",
    "\n",
    "        # build region files to distinct all/easy/hard regions for both maternal and paternal region files\n",
    "        combos = {}\n",
    "        for hap_label, hap_union, hap_easy, hap_hard in [\n",
    "            (\"all\", str(all_regions), None, None),\n",
    "            (\"maternal\", str(mat_union), str(regions['mat_easy']), str(regions['mat_hard'])),\n",
    "            (\"paternal\", str(pat_union), str(regions['pat_easy']), str(regions['pat_hard'])),\n",
    "        ]:\n",
    "            if hap_label == \"all\":\n",
    "                combos[(hap_label, \"all\")] = str(all_regions)\n",
    "                merged_easy = tmpdir/\"both_easy.bed\"\n",
    "                run(f\"cat {regions['mat_easy']} {regions['pat_easy']} | sort -k1,1 -k2,2n | bedtools merge -i - > {merged_easy}\")\n",
    "                merged_hard = tmpdir/\"both_hard.bed\"\n",
    "                run(f\"cat {regions['mat_hard']} {regions['pat_hard']} | sort -k1,1 -k2,2n | bedtools merge -i - > {merged_hard}\")\n",
    "                combos[(hap_label, \"easy\")] = str(merged_easy)\n",
    "                combos[(hap_label, \"hard\")] = str(merged_hard)\n",
    "            else:\n",
    "                combos[(hap_label, \"all\")] = hap_union\n",
    "                combos[(hap_label, \"easy\")] = hap_easy\n",
    "                combos[(hap_label, \"hard\")] = hap_hard\n",
    "\n",
    "        def count_variants_in_region(vcf: str, region_bed: str, variant_type=\"all\"):\n",
    "            \"\"\"\n",
    "            Function to count variants in region with bcftools\n",
    "            variant_type: \"all\", \"snvs\", \"indels\"\n",
    "            \"\"\"\n",
    "\n",
    "            # remove unreliable variants\n",
    "            filtered_vcf = tmpdir / f\"variants_filtered.vcf.gz\"\n",
    "            cmd_filter_unreliable = (\n",
    "                f\"bcftools isec -w1 -C -O z -o {filtered_vcf} \"\n",
    "                f\"{vcf} {UNRELIABLE_GRANDPARENTS}\"\n",
    "            )\n",
    "            subprocess.run(cmd_filter_unreliable, shell=True, check=True)\n",
    "\n",
    "            # index\n",
    "            cmd_index = (\n",
    "                f\"tabix -p vcf -f {filtered_vcf} \"\n",
    "            )\n",
    "            subprocess.run(cmd_index, shell=True, check=True)\n",
    "            \n",
    "            # filter variants of interest\n",
    "            if variant_type == \"all\":\n",
    "                vcf_cmd = f\"bcftools view -R {region_bed} -H {filtered_vcf}\"\n",
    "            elif variant_type.lower() in (\"snv\", \"snvs\", \"snp\", \"snps\"):\n",
    "                vcf_cmd = f\"bcftools view -v snps -R {region_bed} -H {filtered_vcf}\"\n",
    "            elif variant_type.lower() in (\"indel\", \"indels\"):\n",
    "                vcf_cmd = f\"bcftools view -v indels -R {region_bed} -H {filtered_vcf}\"\n",
    "            else:\n",
    "                raise ValueError(\"unknown variant_type: \"+str(variant_type))\n",
    "\n",
    "            # count variants\n",
    "            cmd = vcf_cmd + \" | wc -l\"\n",
    "            out = run(cmd, capture=True)\n",
    "            return int(out)\n",
    "    \n",
    "        # count mutation rates for all combinations of (maternal, paternal, both) x (easy, difficult, all) x (SNVs, indels, either)\n",
    "        rows = []\n",
    "        mutation_type_labels: list[tuple] = [(\"all\",\"all\"), (\"snvs\",\"SNVs\"), (\"indels\",\"Indels\")]\n",
    "        hap_labels: list[str] = [\"all\", \"maternal\", \"paternal\"]\n",
    "        region_types: list[str] = [\"all\",\"easy\",\"hard\"]\n",
    "    \n",
    "        print(\"[INFO] Running counts...\")\n",
    "        for hap in hap_labels:\n",
    "            for mut_key, mut_label in mutation_type_labels:\n",
    "                for reg in region_types:\n",
    "                    region_bed = combos[(hap, reg)]\n",
    "                    try:\n",
    "                        length_bp: int = get_bed_total_length(region_bed)\n",
    "                    except subprocess.CalledProcessError:\n",
    "                        length_bp = 0\n",
    "                    if length_bp == 0:\n",
    "                        count = 0\n",
    "                    else:\n",
    "                        count = count_variants_in_region(variants, region_bed, mut_key)\n",
    "                    per_mb: float = float(count) / (length_bp / 1e6) if length_bp>0 else float(\"nan\")\n",
    "    \n",
    "                    rows.append({\n",
    "                        \"haplotype_scope\": hap,\n",
    "                        \"mutation_type\": mut_label,\n",
    "                        \"region_scope\": reg,\n",
    "                        \"region_bed\": region_bed,\n",
    "                        \"length_bp\": length_bp,\n",
    "                        \"variant_count\": count,\n",
    "                        \"variants_per_Mb\": per_mb,\n",
    "                        \"region_scope_len\": mat_len_total if hap == \"maternal\" else pat_len_total if hap == \"paternal\" else 0\n",
    "                    })\n",
    "                    print(f\"[RESULT] hap={hap} mut={mut_label} reg={reg} count={count} len={length_bp} perMb={per_mb:.3f}\")\n",
    "    \n",
    "        results_df = pd.DataFrame(rows)\n",
    "        results_df = results_df[[\n",
    "            \"haplotype_scope\",\"region_scope\",\"mutation_type\",\"region_bed\",\"length_bp\",\"variant_count\",\"variants_per_Mb\", \"region_scope_len\"\n",
    "        ]]\n",
    "\n",
    "        results_df.to_csv(out_tsv, sep=\"\\t\", index=False)\n",
    "        print(f\"[DONE] summary written to {out_tsv}\")\n",
    "        return results_df\n",
    "    finally:\n",
    "        print(\"[INFO] cleaning tmpdir:\", tmpdir)\n",
    "        shutil.rmtree(tmpdir, ignore_errors=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "cbb76f78-d1a5-42c4-8235-2432ca907931",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_single_chromosome(tmp_root: Path, chrom: str) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Run mutation rate summary for one chromosome\n",
    "    \"\"\"\n",
    "    tmpdir = tmp_root / chrom\n",
    "    tmpdir.mkdir(exist_ok=True)\n",
    "    chr_variants, chr_blocks = filter_inputs_for_chrom(chrom, tmpdir)\n",
    "\n",
    "    df = mk_summary_single(tmpdir, chr_variants, chr_blocks)\n",
    "    df[\"chromosome\"] = chrom\n",
    "    return df"
   ]
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "#### Calculate coarse mutation rates (i.e., without chromosomal differentiation)",
   "id": "34ac9b98b7717229"
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "594b8097-9038-42ca-ae6b-0f089175b337",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[INFO] tmpdir: /tmp/mutrate_zk4371zv\n",
      "[CMD] padding+merge -> /tmp/mutrate_zk4371zv/flagger_ont.padded.bed\n",
      "[CMD] padding+merge -> /tmp/mutrate_zk4371zv/flagger_hifi.padded.bed\n",
      "[CMD] padding+merge -> /tmp/mutrate_zk4371zv/problematic.padded.bed\n",
      "[CMD] combining excludes -> /tmp/mutrate_zk4371zv/combined_exclude.bed\n",
      "[INFO] Running counts (this may take a while depending on VCF size).\n",
      "[RESULT] hap=all mut=all reg=all count=19539 len=2674270000 perMb=7.306\n",
      "[RESULT] hap=all mut=all reg=easy count=355 len=2038533185 perMb=0.174\n",
      "[RESULT] hap=all mut=all reg=hard count=19187 len=635731978 perMb=30.181\n",
      "[RESULT] hap=all mut=SNVs reg=all count=569 len=2674270000 perMb=0.213\n",
      "[RESULT] hap=all mut=SNVs reg=easy count=221 len=2038533185 perMb=0.108\n",
      "[RESULT] hap=all mut=SNVs reg=hard count=348 len=635731978 perMb=0.547\n",
      "[RESULT] hap=all mut=Indels reg=all count=18970 len=2674270000 perMb=7.094\n",
      "[RESULT] hap=all mut=Indels reg=easy count=134 len=2038533185 perMb=0.066\n",
      "[RESULT] hap=all mut=Indels reg=hard count=18839 len=635731978 perMb=29.634\n",
      "[RESULT] hap=maternal mut=all reg=all count=8206 len=1527643660 perMb=5.372\n",
      "[RESULT] hap=maternal mut=all reg=easy count=237 len=1210803285 perMb=0.196\n",
      "[RESULT] hap=maternal mut=all reg=hard count=7970 len=316840375 perMb=25.155\n",
      "[RESULT] hap=maternal mut=SNVs reg=all count=324 len=1527643660 perMb=0.212\n",
      "[RESULT] hap=maternal mut=SNVs reg=easy count=167 len=1210803285 perMb=0.138\n",
      "[RESULT] hap=maternal mut=SNVs reg=hard count=157 len=316840375 perMb=0.496\n",
      "[RESULT] hap=maternal mut=Indels reg=all count=7882 len=1527643660 perMb=5.160\n",
      "[RESULT] hap=maternal mut=Indels reg=easy count=70 len=1210803285 perMb=0.058\n",
      "[RESULT] hap=maternal mut=Indels reg=hard count=7813 len=316840375 perMb=24.659\n",
      "[RESULT] hap=paternal mut=all reg=all count=11333 len=1146621503 perMb=9.884\n",
      "[RESULT] hap=paternal mut=all reg=easy count=118 len=827729900 perMb=0.143\n",
      "[RESULT] hap=paternal mut=all reg=hard count=11217 len=318891603 perMb=35.175\n",
      "[RESULT] hap=paternal mut=SNVs reg=all count=245 len=1146621503 perMb=0.214\n",
      "[RESULT] hap=paternal mut=SNVs reg=easy count=54 len=827729900 perMb=0.065\n",
      "[RESULT] hap=paternal mut=SNVs reg=hard count=191 len=318891603 perMb=0.599\n",
      "[RESULT] hap=paternal mut=Indels reg=all count=11088 len=1146621503 perMb=9.670\n",
      "[RESULT] hap=paternal mut=Indels reg=easy count=64 len=827729900 perMb=0.077\n",
      "[RESULT] hap=paternal mut=Indels reg=hard count=11026 len=318891603 perMb=34.576\n",
      "[DONE] summary written to /tmp/mutrate_zk4371zv/mutation_rates_summary.tsv\n",
      "[INFO] cleaning tmpdir: /tmp/mutrate_zk4371zv\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>haplotype_scope</th>\n",
       "      <th>region_scope</th>\n",
       "      <th>mutation_type</th>\n",
       "      <th>region_bed</th>\n",
       "      <th>length_bp</th>\n",
       "      <th>variant_count</th>\n",
       "      <th>variants_per_Mb</th>\n",
       "      <th>region_scope_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>all</td>\n",
       "      <td>all</td>\n",
       "      <td>all</td>\n",
       "      <td>/tmp/mutrate_zk4371zv/all_regions.bed</td>\n",
       "      <td>2674270000</td>\n",
       "      <td>19539</td>\n",
       "      <td>7.306293</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>all</td>\n",
       "      <td>easy</td>\n",
       "      <td>all</td>\n",
       "      <td>/tmp/mutrate_zk4371zv/both_easy.bed</td>\n",
       "      <td>2038533185</td>\n",
       "      <td>355</td>\n",
       "      <td>0.174145</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>all</td>\n",
       "      <td>hard</td>\n",
       "      <td>all</td>\n",
       "      <td>/tmp/mutrate_zk4371zv/both_hard.bed</td>\n",
       "      <td>635731978</td>\n",
       "      <td>19187</td>\n",
       "      <td>30.180958</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>all</td>\n",
       "      <td>all</td>\n",
       "      <td>SNVs</td>\n",
       "      <td>/tmp/mutrate_zk4371zv/all_regions.bed</td>\n",
       "      <td>2674270000</td>\n",
       "      <td>569</td>\n",
       "      <td>0.212768</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>all</td>\n",
       "      <td>easy</td>\n",
       "      <td>SNVs</td>\n",
       "      <td>/tmp/mutrate_zk4371zv/both_easy.bed</td>\n",
       "      <td>2038533185</td>\n",
       "      <td>221</td>\n",
       "      <td>0.108411</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>all</td>\n",
       "      <td>hard</td>\n",
       "      <td>SNVs</td>\n",
       "      <td>/tmp/mutrate_zk4371zv/both_hard.bed</td>\n",
       "      <td>635731978</td>\n",
       "      <td>348</td>\n",
       "      <td>0.547400</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>all</td>\n",
       "      <td>all</td>\n",
       "      <td>Indels</td>\n",
       "      <td>/tmp/mutrate_zk4371zv/all_regions.bed</td>\n",
       "      <td>2674270000</td>\n",
       "      <td>18970</td>\n",
       "      <td>7.093525</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>all</td>\n",
       "      <td>easy</td>\n",
       "      <td>Indels</td>\n",
       "      <td>/tmp/mutrate_zk4371zv/both_easy.bed</td>\n",
       "      <td>2038533185</td>\n",
       "      <td>134</td>\n",
       "      <td>0.065734</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>all</td>\n",
       "      <td>hard</td>\n",
       "      <td>Indels</td>\n",
       "      <td>/tmp/mutrate_zk4371zv/both_hard.bed</td>\n",
       "      <td>635731978</td>\n",
       "      <td>18839</td>\n",
       "      <td>29.633557</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>maternal</td>\n",
       "      <td>all</td>\n",
       "      <td>all</td>\n",
       "      <td>/tmp/mutrate_zk4371zv/mat_union.bed</td>\n",
       "      <td>1527643660</td>\n",
       "      <td>8206</td>\n",
       "      <td>5.371672</td>\n",
       "      <td>1527643660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>maternal</td>\n",
       "      <td>easy</td>\n",
       "      <td>all</td>\n",
       "      <td>/tmp/mutrate_zk4371zv/mat_easy.selected.bed</td>\n",
       "      <td>1210803285</td>\n",
       "      <td>237</td>\n",
       "      <td>0.195738</td>\n",
       "      <td>1527643660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>maternal</td>\n",
       "      <td>hard</td>\n",
       "      <td>all</td>\n",
       "      <td>/tmp/mutrate_zk4371zv/mat_hard.selected.bed</td>\n",
       "      <td>316840375</td>\n",
       "      <td>7970</td>\n",
       "      <td>25.154622</td>\n",
       "      <td>1527643660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>maternal</td>\n",
       "      <td>all</td>\n",
       "      <td>SNVs</td>\n",
       "      <td>/tmp/mutrate_zk4371zv/mat_union.bed</td>\n",
       "      <td>1527643660</td>\n",
       "      <td>324</td>\n",
       "      <td>0.212091</td>\n",
       "      <td>1527643660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>maternal</td>\n",
       "      <td>easy</td>\n",
       "      <td>SNVs</td>\n",
       "      <td>/tmp/mutrate_zk4371zv/mat_easy.selected.bed</td>\n",
       "      <td>1210803285</td>\n",
       "      <td>167</td>\n",
       "      <td>0.137925</td>\n",
       "      <td>1527643660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>maternal</td>\n",
       "      <td>hard</td>\n",
       "      <td>SNVs</td>\n",
       "      <td>/tmp/mutrate_zk4371zv/mat_hard.selected.bed</td>\n",
       "      <td>316840375</td>\n",
       "      <td>157</td>\n",
       "      <td>0.495518</td>\n",
       "      <td>1527643660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>maternal</td>\n",
       "      <td>all</td>\n",
       "      <td>Indels</td>\n",
       "      <td>/tmp/mutrate_zk4371zv/mat_union.bed</td>\n",
       "      <td>1527643660</td>\n",
       "      <td>7882</td>\n",
       "      <td>5.159580</td>\n",
       "      <td>1527643660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>maternal</td>\n",
       "      <td>easy</td>\n",
       "      <td>Indels</td>\n",
       "      <td>/tmp/mutrate_zk4371zv/mat_easy.selected.bed</td>\n",
       "      <td>1210803285</td>\n",
       "      <td>70</td>\n",
       "      <td>0.057813</td>\n",
       "      <td>1527643660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>maternal</td>\n",
       "      <td>hard</td>\n",
       "      <td>Indels</td>\n",
       "      <td>/tmp/mutrate_zk4371zv/mat_hard.selected.bed</td>\n",
       "      <td>316840375</td>\n",
       "      <td>7813</td>\n",
       "      <td>24.659105</td>\n",
       "      <td>1527643660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>paternal</td>\n",
       "      <td>all</td>\n",
       "      <td>all</td>\n",
       "      <td>/tmp/mutrate_zk4371zv/pat_union.bed</td>\n",
       "      <td>1146621503</td>\n",
       "      <td>11333</td>\n",
       "      <td>9.883820</td>\n",
       "      <td>1146621503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>paternal</td>\n",
       "      <td>easy</td>\n",
       "      <td>all</td>\n",
       "      <td>/tmp/mutrate_zk4371zv/pat_easy.selected.bed</td>\n",
       "      <td>827729900</td>\n",
       "      <td>118</td>\n",
       "      <td>0.142559</td>\n",
       "      <td>1146621503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>paternal</td>\n",
       "      <td>hard</td>\n",
       "      <td>all</td>\n",
       "      <td>/tmp/mutrate_zk4371zv/pat_hard.selected.bed</td>\n",
       "      <td>318891603</td>\n",
       "      <td>11217</td>\n",
       "      <td>35.174962</td>\n",
       "      <td>1146621503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>paternal</td>\n",
       "      <td>all</td>\n",
       "      <td>SNVs</td>\n",
       "      <td>/tmp/mutrate_zk4371zv/pat_union.bed</td>\n",
       "      <td>1146621503</td>\n",
       "      <td>245</td>\n",
       "      <td>0.213671</td>\n",
       "      <td>1146621503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>paternal</td>\n",
       "      <td>easy</td>\n",
       "      <td>SNVs</td>\n",
       "      <td>/tmp/mutrate_zk4371zv/pat_easy.selected.bed</td>\n",
       "      <td>827729900</td>\n",
       "      <td>54</td>\n",
       "      <td>0.065239</td>\n",
       "      <td>1146621503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>paternal</td>\n",
       "      <td>hard</td>\n",
       "      <td>SNVs</td>\n",
       "      <td>/tmp/mutrate_zk4371zv/pat_hard.selected.bed</td>\n",
       "      <td>318891603</td>\n",
       "      <td>191</td>\n",
       "      <td>0.598950</td>\n",
       "      <td>1146621503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>paternal</td>\n",
       "      <td>all</td>\n",
       "      <td>Indels</td>\n",
       "      <td>/tmp/mutrate_zk4371zv/pat_union.bed</td>\n",
       "      <td>1146621503</td>\n",
       "      <td>11088</td>\n",
       "      <td>9.670148</td>\n",
       "      <td>1146621503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>paternal</td>\n",
       "      <td>easy</td>\n",
       "      <td>Indels</td>\n",
       "      <td>/tmp/mutrate_zk4371zv/pat_easy.selected.bed</td>\n",
       "      <td>827729900</td>\n",
       "      <td>64</td>\n",
       "      <td>0.077320</td>\n",
       "      <td>1146621503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>paternal</td>\n",
       "      <td>hard</td>\n",
       "      <td>Indels</td>\n",
       "      <td>/tmp/mutrate_zk4371zv/pat_hard.selected.bed</td>\n",
       "      <td>318891603</td>\n",
       "      <td>11026</td>\n",
       "      <td>34.576012</td>\n",
       "      <td>1146621503</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   haplotype_scope region_scope mutation_type  \\\n",
       "0              all          all           all   \n",
       "1              all         easy           all   \n",
       "2              all         hard           all   \n",
       "3              all          all          SNVs   \n",
       "4              all         easy          SNVs   \n",
       "5              all         hard          SNVs   \n",
       "6              all          all        Indels   \n",
       "7              all         easy        Indels   \n",
       "8              all         hard        Indels   \n",
       "9         maternal          all           all   \n",
       "10        maternal         easy           all   \n",
       "11        maternal         hard           all   \n",
       "12        maternal          all          SNVs   \n",
       "13        maternal         easy          SNVs   \n",
       "14        maternal         hard          SNVs   \n",
       "15        maternal          all        Indels   \n",
       "16        maternal         easy        Indels   \n",
       "17        maternal         hard        Indels   \n",
       "18        paternal          all           all   \n",
       "19        paternal         easy           all   \n",
       "20        paternal         hard           all   \n",
       "21        paternal          all          SNVs   \n",
       "22        paternal         easy          SNVs   \n",
       "23        paternal         hard          SNVs   \n",
       "24        paternal          all        Indels   \n",
       "25        paternal         easy        Indels   \n",
       "26        paternal         hard        Indels   \n",
       "\n",
       "                                     region_bed   length_bp  variant_count  \\\n",
       "0         /tmp/mutrate_zk4371zv/all_regions.bed  2674270000          19539   \n",
       "1           /tmp/mutrate_zk4371zv/both_easy.bed  2038533185            355   \n",
       "2           /tmp/mutrate_zk4371zv/both_hard.bed   635731978          19187   \n",
       "3         /tmp/mutrate_zk4371zv/all_regions.bed  2674270000            569   \n",
       "4           /tmp/mutrate_zk4371zv/both_easy.bed  2038533185            221   \n",
       "5           /tmp/mutrate_zk4371zv/both_hard.bed   635731978            348   \n",
       "6         /tmp/mutrate_zk4371zv/all_regions.bed  2674270000          18970   \n",
       "7           /tmp/mutrate_zk4371zv/both_easy.bed  2038533185            134   \n",
       "8           /tmp/mutrate_zk4371zv/both_hard.bed   635731978          18839   \n",
       "9           /tmp/mutrate_zk4371zv/mat_union.bed  1527643660           8206   \n",
       "10  /tmp/mutrate_zk4371zv/mat_easy.selected.bed  1210803285            237   \n",
       "11  /tmp/mutrate_zk4371zv/mat_hard.selected.bed   316840375           7970   \n",
       "12          /tmp/mutrate_zk4371zv/mat_union.bed  1527643660            324   \n",
       "13  /tmp/mutrate_zk4371zv/mat_easy.selected.bed  1210803285            167   \n",
       "14  /tmp/mutrate_zk4371zv/mat_hard.selected.bed   316840375            157   \n",
       "15          /tmp/mutrate_zk4371zv/mat_union.bed  1527643660           7882   \n",
       "16  /tmp/mutrate_zk4371zv/mat_easy.selected.bed  1210803285             70   \n",
       "17  /tmp/mutrate_zk4371zv/mat_hard.selected.bed   316840375           7813   \n",
       "18          /tmp/mutrate_zk4371zv/pat_union.bed  1146621503          11333   \n",
       "19  /tmp/mutrate_zk4371zv/pat_easy.selected.bed   827729900            118   \n",
       "20  /tmp/mutrate_zk4371zv/pat_hard.selected.bed   318891603          11217   \n",
       "21          /tmp/mutrate_zk4371zv/pat_union.bed  1146621503            245   \n",
       "22  /tmp/mutrate_zk4371zv/pat_easy.selected.bed   827729900             54   \n",
       "23  /tmp/mutrate_zk4371zv/pat_hard.selected.bed   318891603            191   \n",
       "24          /tmp/mutrate_zk4371zv/pat_union.bed  1146621503          11088   \n",
       "25  /tmp/mutrate_zk4371zv/pat_easy.selected.bed   827729900             64   \n",
       "26  /tmp/mutrate_zk4371zv/pat_hard.selected.bed   318891603          11026   \n",
       "\n",
       "    variants_per_Mb  region_scope_len  \n",
       "0          7.306293                 0  \n",
       "1          0.174145                 0  \n",
       "2         30.180958                 0  \n",
       "3          0.212768                 0  \n",
       "4          0.108411                 0  \n",
       "5          0.547400                 0  \n",
       "6          7.093525                 0  \n",
       "7          0.065734                 0  \n",
       "8         29.633557                 0  \n",
       "9          5.371672        1527643660  \n",
       "10         0.195738        1527643660  \n",
       "11        25.154622        1527643660  \n",
       "12         0.212091        1527643660  \n",
       "13         0.137925        1527643660  \n",
       "14         0.495518        1527643660  \n",
       "15         5.159580        1527643660  \n",
       "16         0.057813        1527643660  \n",
       "17        24.659105        1527643660  \n",
       "18         9.883820        1146621503  \n",
       "19         0.142559        1146621503  \n",
       "20        35.174962        1146621503  \n",
       "21         0.213671        1146621503  \n",
       "22         0.065239        1146621503  \n",
       "23         0.598950        1146621503  \n",
       "24         9.670148        1146621503  \n",
       "25         0.077320        1146621503  \n",
       "26        34.576012        1146621503  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "base_tmp: str = os.environ.get(\"SCRATCH\", None) or os.environ.get(\"TMPDIR\", None) or \"/tmp\"\n",
    "tmpdir: Path = Path(tempfile.mkdtemp(prefix=\"mutrate_\", dir=base_tmp))\n",
    "print(\"[INFO] tmpdir:\", tmpdir)\n",
    "\n",
    "mk_summary_single(tmpdir, variants_vcf, all_blocks)"
   ]
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "#### Calculate fine mutation rates (i.e., with chromosomal differentiation)",
   "id": "14f2a43a3c25050a"
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "6ae83805-7bfd-4e8b-ac0d-f0c7c03da09d",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true,
     "source_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[INFO] tmpdir: /tmp/mutrate_896r0tio\n",
      "[CMD] padding+merge ->[CMD] padding+merge ->  /tmp/mutrate_896r0tio/chr16/flagger_ont.padded.bed/tmp/mutrate_896r0tio/chr12/flagger_ont.padded.bed\n",
      "\n",
      "[CMD] padding+merge -> /tmp/mutrate_896r0tio/chr13/flagger_ont.padded.bed\n",
      "[CMD] padding+merge ->[CMD] padding+merge ->[CMD] padding+merge ->[CMD] padding+merge ->[CMD] padding+merge ->[CMD] padding+merge ->  [CMD] padding+merge ->[CMD] padding+merge ->[CMD] padding+merge -> [CMD] padding+merge -> [CMD] padding+merge -> [CMD] padding+merge ->/tmp/mutrate_896r0tio/chr6/flagger_ont.padded.bed[CMD] padding+merge ->/tmp/mutrate_896r0tio/chr5/flagger_ont.padded.bed    /tmp/mutrate_896r0tio/chr10/flagger_ont.padded.bed /tmp/mutrate_896r0tio/chr8/flagger_ont.padded.bed/tmp/mutrate_896r0tio/chr11/flagger_ont.padded.bed \n",
      " \n",
      " /tmp/mutrate_896r0tio/chr9/flagger_ont.padded.bed/tmp/mutrate_896r0tio/chr2/flagger_ont.padded.bed/tmp/mutrate_896r0tio/chr4/flagger_ont.padded.bed/tmp/mutrate_896r0tio/chr15/flagger_ont.padded.bed\n",
      "/tmp/mutrate_896r0tio/chr1/flagger_ont.padded.bed\n",
      "\n",
      "/tmp/mutrate_896r0tio/chr3/flagger_ont.padded.bed/tmp/mutrate_896r0tio/chr14/flagger_ont.padded.bed/tmp/mutrate_896r0tio/chr7/flagger_ont.padded.bed\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "[CMD] padding+merge ->[CMD] padding+merge ->[CMD] padding+merge ->   /tmp/mutrate_896r0tio/chr16/flagger_hifi.padded.bed/tmp/mutrate_896r0tio/chr13/flagger_hifi.padded.bed/tmp/mutrate_896r0tio/chr12/flagger_hifi.padded.bed\n",
      "\n",
      "\n",
      "[CMD] padding+merge -> [CMD] padding+merge ->[CMD] padding+merge ->/tmp/mutrate_896r0tio/chr6/flagger_hifi.padded.bed[CMD] padding+merge -> [CMD] padding+merge -> \n",
      " /tmp/mutrate_896r0tio/chr10/flagger_hifi.padded.bed[CMD] padding+merge ->/tmp/mutrate_896r0tio/chr5/flagger_hifi.padded.bed[CMD] padding+merge -> /tmp/mutrate_896r0tio/chr8/flagger_hifi.padded.bed[CMD] padding+merge ->\n",
      "[CMD] padding+merge ->[CMD] padding+merge ->[CMD] padding+merge -> \n",
      " [CMD] padding+merge ->/tmp/mutrate_896r0tio/chr11/flagger_hifi.padded.bed\n",
      "   [CMD] padding+merge -> /tmp/mutrate_896r0tio/chr9/flagger_hifi.padded.bed[CMD] padding+merge ->/tmp/mutrate_896r0tio/chr2/flagger_hifi.padded.bed \n",
      "[CMD] padding+merge ->[CMD] padding+merge ->/tmp/mutrate_896r0tio/chr15/flagger_hifi.padded.bed/tmp/mutrate_896r0tio/chr14/flagger_hifi.padded.bed/tmp/mutrate_896r0tio/chr1/flagger_hifi.padded.bed \n",
      "/tmp/mutrate_896r0tio/chr3/flagger_hifi.padded.bed \n",
      "/tmp/mutrate_896r0tio/chr4/flagger_hifi.padded.bed  \n",
      "\n",
      "\n",
      "/tmp/mutrate_896r0tio/chr7/flagger_hifi.padded.bed\n",
      "/tmp/mutrate_896r0tio/chr12/problematic.padded.bed\n",
      "/tmp/mutrate_896r0tio/chr16/problematic.padded.bed/tmp/mutrate_896r0tio/chr13/problematic.padded.bed\n",
      "\n",
      "\n",
      "\n",
      "[CMD] padding+merge -> /tmp/mutrate_896r0tio/chr6/problematic.padded.bed\n",
      "[CMD] padding+merge -> [CMD] padding+merge ->/tmp/mutrate_896r0tio/chr10/problematic.padded.bed[CMD] padding+merge -> \n",
      " /tmp/mutrate_896r0tio/chr5/problematic.padded.bed/tmp/mutrate_896r0tio/chr8/problematic.padded.bed\n",
      "\n",
      "[CMD] padding+merge ->[CMD] padding+merge -> /tmp/mutrate_896r0tio/chr11/problematic.padded.bed [CMD] padding+merge ->/tmp/mutrate_896r0tio/chr9/problematic.padded.bed[CMD] padding+merge ->\n",
      "[CMD] padding+merge ->\n",
      "   [CMD] padding+merge ->[CMD] padding+merge ->/tmp/mutrate_896r0tio/chr14/problematic.padded.bed  /tmp/mutrate_896r0tio/chr2/problematic.padded.bed/tmp/mutrate_896r0tio/chr4/problematic.padded.bed\n",
      "[CMD] padding+merge ->/tmp/mutrate_896r0tio/chr15/problematic.padded.bed/tmp/mutrate_896r0tio/chr3/problematic.padded.bed[CMD] padding+merge ->\n",
      "\n",
      "[CMD] combining excludes ->  \n",
      "\n",
      " /tmp/mutrate_896r0tio/chr7/problematic.padded.bed/tmp/mutrate_896r0tio/chr1/problematic.padded.bed/tmp/mutrate_896r0tio/chr16/combined_exclude.bed\n",
      "\n",
      "[CMD] combining excludes ->\n",
      " [CMD] combining excludes ->/tmp/mutrate_896r0tio/chr12/combined_exclude.bed[CMD] combining excludes -> [CMD] combining excludes ->\n",
      " /tmp/mutrate_896r0tio/chr9/combined_exclude.bed/tmp/mutrate_896r0tio/chr5/combined_exclude.bed \n",
      "[CMD] combining excludes ->\n",
      "/tmp/mutrate_896r0tio/chr10/combined_exclude.bed[CMD] combining excludes ->  \n",
      "/tmp/mutrate_896r0tio/chr6/combined_exclude.bed/tmp/mutrate_896r0tio/chr13/combined_exclude.bed\n",
      "[CMD] combining excludes ->\n",
      " /tmp/mutrate_896r0tio/chr8/combined_exclude.bed\n",
      "[CMD] combining excludes -> /tmp/mutrate_896r0tio/chr11/combined_exclude.bed\n",
      "[CMD] combining excludes ->[CMD] combining excludes ->[CMD] combining excludes ->[CMD] combining excludes ->[CMD] combining excludes ->   [CMD] combining excludes ->  [CMD] combining excludes ->/tmp/mutrate_896r0tio/chr14/combined_exclude.bed/tmp/mutrate_896r0tio/chr2/combined_exclude.bed/tmp/mutrate_896r0tio/chr4/combined_exclude.bed/tmp/mutrate_896r0tio/chr15/combined_exclude.bed /tmp/mutrate_896r0tio/chr7/combined_exclude.bed\n",
      " \n",
      "\n",
      "/tmp/mutrate_896r0tio/chr3/combined_exclude.bed\n",
      "\n",
      "/tmp/mutrate_896r0tio/chr1/combined_exclude.bed\n",
      "\n",
      "[INFO] Running counts (this may take a while depending on VCF size).\n",
      "[RESULT] hap=all mut=all reg=all count=812 len=75751317 perMb=10.719\n",
      "[INFO] Running counts (this may take a while depending on VCF size).\n",
      "[RESULT] hap=all mut=all reg=all count=824 len=87138247 perMb=9.456\n",
      "[INFO] Running counts (this may take a while depending on VCF size).\n",
      "[INFO] Running counts (this may take a while depending on VCF size).\n",
      "[RESULT] hap=all mut=all reg=all count=716 len=97131486 perMb=7.371\n",
      "[RESULT] hap=all mut=all reg=all count=518 len=87672034 perMb=5.908\n",
      "[INFO] Running counts (this may take a while depending on VCF size).\n",
      "[RESULT] hap=all mut=all reg=easy count=3 len=48527966 perMb=0.062\n",
      "[RESULT] hap=all mut=all reg=all count=973 len=125620867 perMb=7.746\n",
      "[INFO] Running counts (this may take a while depending on VCF size).\n",
      "[RESULT] hap=all mut=all reg=easy count=6 len=55574766 perMb=0.108\n",
      "[RESULT] hap=all mut=all reg=all count=733 len=116426940 perMb=6.296\n",
      "[INFO] Running counts (this may take a while depending on VCF size).\n",
      "[RESULT] hap=all mut=all reg=all count=809 len=130449725 perMb=6.202\n",
      "[RESULT] hap=all mut=all reg=easy count=10 len=70620705 perMb=0.142\n",
      "[RESULT] hap=all mut=all reg=easy count=17 len=63539435 perMb=0.268\n",
      "[INFO] Running counts (this may take a while depending on VCF size).\n",
      "[RESULT] hap=all mut=all reg=hard count=809 len=27223351 perMb=29.717\n",
      "[RESULT] hap=all mut=all reg=all count=949 len=138492834 perMb=6.852\n",
      "[RESULT] hap=all mut=SNVs reg=all count=19 len=75751317 perMb=0.251\n",
      "[RESULT] hap=all mut=all reg=easy count=10 len=82768334 perMb=0.121\n",
      "[INFO] Running counts (this may take a while depending on VCF size).[RESULT] hap=all mut=all reg=hard count=818 len=31563481 perMb=25.916\n",
      "\n",
      "[INFO] Running counts (this may take a while depending on VCF size).\n",
      "[RESULT] hap=all mut=all reg=easy count=22 len=92911699 perMb=0.237\n",
      "[RESULT] hap=all mut=SNVs reg=all count=19 len=87138247 perMb=0.218[RESULT] hap=all mut=all reg=all count=1093 len=137502700 perMb=7.949\n",
      "\n",
      "[RESULT] hap=all mut=all reg=all count=1103 len=126695273 perMb=8.706\n",
      "[INFO] Running counts (this may take a while depending on VCF size).\n",
      "[INFO] Running counts (this may take a while depending on VCF size).\n",
      "[INFO] Running counts (this may take a while depending on VCF size).\n",
      "[RESULT] hap=all mut=all reg=all count=1183 len=156673292 perMb=7.551\n",
      "[RESULT] hap=all mut=all reg=all count=1002 len=160297126 perMb=6.251\n",
      "[INFO] Running counts (this may take a while depending on VCF size).\n",
      "[RESULT] hap=all mut=all reg=all count=894 len=156635364 perMb=5.708\n",
      "[RESULT] hap=all mut=all reg=easy count=22 len=103594563 perMb=0.212\n",
      "[RESULT] hap=all mut=all reg=all count=827 len=154890228 perMb=5.339\n",
      "[RESULT] hap=all mut=all reg=hard count=706 len=26510781 perMb=26.631\n",
      "[RESULT] hap=all mut=all reg=hard count=501 len=24132599 perMb=20.760\n",
      "[RESULT] hap=all mut=SNVs reg=all count=15 len=97131486 perMb=0.154\n",
      "[RESULT] hap=all mut=SNVs reg=all count=19 len=87672034 perMb=0.217\n",
      "[RESULT] hap=all mut=SNVs reg=easy count=1 len=48527966 perMb=0.021\n",
      "[RESULT] hap=all mut=all reg=hard count=963 len=42852533 perMb=22.472\n",
      "[RESULT] hap=all mut=SNVs reg=easy count=2 len=55574766 perMb=0.036\n",
      "[RESULT] hap=all mut=SNVs reg=all count=21 len=125620867 perMb=0.167\n",
      "[RESULT] hap=all mut=all reg=easy count=16 len=105437901 perMb=0.152\n",
      "[RESULT] hap=all mut=all reg=hard count=711 len=23515241 perMb=30.236\n",
      "[RESULT] hap=all mut=SNVs reg=all count=38 len=116426940 perMb=0.326\n",
      "[RESULT] hap=all mut=all reg=hard count=787 len=26855162 perMb=29.305\n",
      "[RESULT] hap=all mut=all reg=easy count=13 len=126508093 perMb=0.103\n",
      "[RESULT] hap=all mut=SNVs reg=all count=18 len=130449725 perMb=0.138\n",
      "[INFO] Running counts (this may take a while depending on VCF size).\n",
      "[RESULT] hap=all mut=SNVs reg=easy count=3 len=70620705 perMb=0.042\n",
      "[RESULT] hap=all mut=SNVs reg=easy count=14 len=63539435 perMb=0.220\n",
      "[RESULT] hap=all mut=all reg=easy count=22 len=127839355 perMb=0.172\n",
      "[RESULT] hap=all mut=SNVs reg=hard count=18 len=27223351 perMb=0.661[RESULT] hap=all mut=all reg=easy count=35 len=103592288 perMb=0.338\n",
      "\n",
      "[RESULT] hap=all mut=all reg=all count=1945 len=227278701 perMb=8.558\n",
      "[INFO] Running counts (this may take a while depending on VCF size).\n",
      "[RESULT] hap=all mut=Indels reg=all count=793 len=75751317 perMb=10.468\n",
      "[RESULT] hap=all mut=all reg=easy count=10 len=128573090 perMb=0.078\n",
      "[RESULT] hap=all mut=all reg=easy count=5 len=112555838 perMb=0.044\n",
      "[RESULT] hap=all mut=all reg=all count=1546 len=228722703 perMb=6.759\n",
      "[RESULT] hap=all mut=SNVs reg=hard count=17 len=31563481 perMb=0.539\n",
      "[RESULT] hap=all mut=all reg=easy count=7 len=128597278 perMb=0.054\n",
      "[RESULT] hap=all mut=Indels reg=all count=805 len=87138247 perMb=9.238\n",
      "[RESULT] hap=all mut=SNVs reg=easy count=2 len=82768334 perMb=0.024\n",
      "[RESULT] hap=all mut=SNVs reg=easy count=19 len=92911699 perMb=0.204\n",
      "[RESULT] hap=all mut=all reg=hard count=933 len=33054933 perMb=28.226\n",
      "[RESULT] hap=all mut=SNVs reg=all count=28 len=138492834 perMb=0.202\n",
      "[RESULT] hap=all mut=SNVs reg=hard count=12 len=26510781 perMb=0.453[RESULT] hap=all mut=SNVs reg=hard count=5 len=24132599 perMb=0.207\n",
      "\n",
      "[RESULT] hap=all mut=SNVs reg=easy count=11 len=103594563 perMb=0.106\n",
      "[RESULT] hap=all mut=Indels reg=all count=499 len=87672034 perMb=5.692\n",
      "[RESULT] hap=all mut=Indels reg=all count=701 len=97131486 perMb=7.217\n",
      "[RESULT] hap=all mut=Indels reg=easy count=2 len=48527966 perMb=0.041\n",
      "[RESULT] hap=all mut=all reg=hard count=814 len=28382135 perMb=28.680\n",
      "[RESULT] hap=all mut=SNVs reg=all count=30 len=154890228 perMb=0.194\n",
      "[RESULT] hap=all mut=Indels reg=easy count=4 len=55574766 perMb=0.072\n",
      "[RESULT] hap=all mut=SNVs reg=hard count=19 len=42852533 perMb=0.443\n",
      "[RESULT] hap=all mut=SNVs reg=hard count=19 len=23515241 perMb=0.808\n",
      "[RESULT] hap=all mut=all reg=hard count=980 len=32457771 perMb=30.193\n",
      "[RESULT] hap=all mut=all reg=hard count=884 len=28062274 perMb=31.501\n",
      "[RESULT] hap=all mut=all reg=hard count=1068 len=23102985 perMb=46.228\n",
      "[RESULT] hap=all mut=Indels reg=all count=952 len=125620867 perMb=7.578\n",
      "[RESULT] hap=all mut=Indels reg=all count=695 len=116426940 perMb=5.969\n",
      "[RESULT] hap=all mut=SNVs reg=all count=23 len=160297126 perMb=0.143\n",
      "[RESULT] hap=all mut=SNVs reg=all count=16 len=156635364 perMb=0.102\n",
      "[RESULT] hap=all mut=SNVs reg=all count=38 len=126695273 perMb=0.300\n",
      "[RESULT] hap=all mut=Indels reg=easy count=7 len=70620705 perMb=0.099[RESULT] hap=all mut=SNVs reg=hard count=7 len=26855162 perMb=0.261\n",
      "\n",
      "[RESULT] hap=all mut=Indels reg=hard count=791 len=27223351 perMb=29.056\n",
      "[RESULT] hap=all mut=all reg=hard count=1088 len=24946862 perMb=43.613[RESULT] hap=all mut=Indels reg=easy count=3 len=63539435 perMb=0.047\n",
      "\n",
      "[RESULT] hap=all mut=all reg=hard count=1177 len=28076014 perMb=41.922\n",
      "[RESULT] hap=all mut=Indels reg=all count=791 len=130449725 perMb=6.064\n",
      "[RESULT] hap=maternal mut=all reg=all count=171 len=21622535 perMb=7.908\n",
      "[RESULT] hap=all mut=SNVs reg=easy count=8 len=105437901 perMb=0.076\n",
      "[RESULT] hap=all mut=Indels reg=hard count=801 len=31563481 perMb=25.377\n",
      "[RESULT] hap=all mut=SNVs reg=all count=19 len=137502700 perMb=0.138\n",
      "[RESULT] hap=all mut=SNVs reg=all count=16 len=156673292 perMb=0.102\n",
      "[RESULT] hap=maternal mut=all reg=all count=144 len=21927748 perMb=6.567\n",
      "[RESULT] hap=all mut=SNVs reg=easy count=5 len=126508093 perMb=0.040\n",
      "[RESULT] hap=all mut=all reg=easy count=18 len=184514589 perMb=0.098\n",
      "[RESULT] hap=maternal mut=all reg=easy count=0 len=17517058 perMb=0.000\n",
      "[RESULT] hap=all mut=Indels reg=easy count=3 len=92911699 perMb=0.032\n",
      "[RESULT] hap=all mut=Indels reg=easy count=8 len=82768334 perMb=0.097\n",
      "[RESULT] hap=maternal mut=all reg=easy count=0 len=18533062 perMb=0.000\n",
      "[RESULT] hap=maternal mut=all reg=hard count=171 len=4105477 perMb=41.652\n",
      "[RESULT] hap=maternal mut=SNVs reg=all count=6 len=21622535 perMb=0.277\n",
      "[RESULT] hap=all mut=Indels reg=hard count=694 len=26510781 perMb=26.178\n",
      "[RESULT] hap=all mut=SNVs reg=easy count=5 len=128573090 perMb=0.039\n",
      "[RESULT] hap=all mut=SNVs reg=easy count=14 len=127839355 perMb=0.110\n",
      "[RESULT] hap=maternal mut=all reg=hard count=144 len=3394686 perMb=42.419\n",
      "[RESULT] hap=all mut=Indels reg=hard count=496 len=24132599 perMb=20.553\n",
      "[RESULT] hap=maternal mut=all reg=all count=220 len=42603212 perMb=5.164\n",
      "[RESULT] hap=all mut=Indels reg=easy count=11 len=103594563 perMb=0.106\n",
      "[RESULT] hap=maternal mut=all reg=all count=125 len=36338486 perMb=3.440\n",
      "[RESULT] hap=maternal mut=SNVs reg=all count=2 len=21927748 perMb=0.091\n",
      "[RESULT] hap=all mut=SNVs reg=easy count=24 len=103592288 perMb=0.232\n",
      "[RESULT] hap=all mut=all reg=easy count=36 len=170403415 perMb=0.211\n",
      "[RESULT] hap=maternal mut=SNVs reg=easy count=0 len=17517058 perMb=0.000\n",
      "[RESULT] hap=all mut=SNVs reg=hard count=25 len=28382135 perMb=0.881\n",
      "[RESULT] hap=all mut=SNVs reg=hard count=20 len=33054933 perMb=0.605\n",
      "[RESULT] hap=all mut=Indels reg=hard count=944 len=42852533 perMb=22.029\n",
      "[RESULT] hap=maternal mut=all reg=easy count=3 len=21392665 perMb=0.140\n",
      "[RESULT] hap=all mut=Indels reg=all count=797 len=154890228 perMb=5.146\n",
      "[RESULT] hap=all mut=Indels reg=all count=921 len=138492834 perMb=6.650\n",
      "[RESULT] hap=maternal mut=SNVs reg=easy count=0 len=18533062 perMb=0.000\n",
      "[RESULT] hap=maternal mut=all reg=all count=231 len=46332509 perMb=4.986\n",
      "[RESULT] hap=maternal mut=all reg=easy count=4 len=37353745 perMb=0.107\n",
      "[RESULT] hap=maternal mut=SNVs reg=hard count=6 len=4105477 perMb=1.461\n",
      "[RESULT] hap=all mut=Indels reg=hard count=692 len=23515241 perMb=29.428\n",
      "[RESULT] hap=maternal mut=Indels reg=all count=165 len=21622535 perMb=7.631[RESULT] hap=maternal mut=all reg=all count=367 len=77297733 perMb=4.748\n",
      "\n",
      "[RESULT] hap=all mut=SNVs reg=easy count=3 len=112555838 perMb=0.027\n",
      "[RESULT] hap=maternal mut=all reg=hard count=122 len=14945821 perMb=8.163\n",
      "[RESULT] hap=maternal mut=SNVs reg=all count=6 len=36338486 perMb=0.165\n",
      "[RESULT] hap=maternal mut=SNVs reg=hard count=2 len=3394686 perMb=0.589\n",
      "[RESULT] hap=maternal mut=all reg=easy count=1 len=39088271 perMb=0.026\n",
      "[RESULT] hap=all mut=SNVs reg=easy count=1 len=128597278 perMb=0.008\n",
      "[RESULT] hap=maternal mut=all reg=hard count=216 len=5249467 perMb=41.147[RESULT] hap=maternal mut=Indels reg=all count=142 len=21927748 perMb=6.476\n",
      "\n",
      "[RESULT] hap=maternal mut=Indels reg=easy count=0 len=17517058 perMb=0.000\n",
      "[RESULT] hap=all mut=Indels reg=hard count=780 len=26855162 perMb=29.045\n",
      "[RESULT] hap=maternal mut=SNVs reg=all count=3 len=42603212 perMb=0.070\n",
      "[RESULT] hap=all mut=SNVs reg=hard count=11 len=28062274 perMb=0.392\n",
      "[RESULT] hap=maternal mut=SNVs reg=easy count=3 len=21392665 perMb=0.140\n",
      "[RESULT] hap=maternal mut=all reg=all count=595 len=110274742 perMb=5.396\n",
      "[RESULT] hap=all mut=Indels reg=all count=878 len=156635364 perMb=5.605\n",
      "[RESULT] hap=maternal mut=all reg=easy count=19 len=65589548 perMb=0.290\n",
      "[RESULT] hap=maternal mut=Indels reg=easy count=0 len=18533062 perMb=0.000\n",
      "[RESULT] hap=maternal mut=all reg=hard count=230 len=7244238 perMb=31.749\n",
      "[RESULT] hap=maternal mut=Indels reg=hard count=165 len=4105477 perMb=40.190\n",
      "[RESULT] hap=all mut=SNVs reg=hard count=9 len=32457771 perMb=0.277\n",
      "[RESULT] hap=all mut=all reg=hard count=1528 len=44208114 perMb=34.564\n",
      "[RESULT] hap=paternal mut=all reg=all count=641 len=54128782 perMb=11.842\n",
      "[RESULT] hap=maternal mut=SNVs reg=all count=6 len=46332509 perMb=0.129\n",
      "[RESULT] hap=maternal mut=SNVs reg=hard count=3 len=14945821 perMb=0.201\n",
      "[RESULT] hap=all mut=Indels reg=all count=979 len=160297126 perMb=6.107\n",
      "[RESULT] hap=maternal mut=SNVs reg=easy count=2 len=37353745 perMb=0.054\n",
      "[RESULT] hap=all mut=SNVs reg=all count=22 len=228722703 perMb=0.096\n",
      "[RESULT] hap=maternal mut=Indels reg=all count=119 len=36338486 perMb=3.275\n",
      "[RESULT] hap=all mut=Indels reg=easy count=8 len=126508093 perMb=0.063\n",
      "[RESULT] hap=maternal mut=Indels reg=hard count=142 len=3394686 perMb=41.830\n",
      "[RESULT] hap=all mut=Indels reg=easy count=8 len=105437901 perMb=0.076\n",
      "[RESULT] hap=paternal mut=all reg=all count=680 len=65210499 perMb=10.428\n",
      "[RESULT] hap=all mut=SNVs reg=hard count=14 len=23102985 perMb=0.606\n",
      "[RESULT] hap=maternal mut=SNVs reg=easy count=1 len=39088271 perMb=0.026\n",
      "[RESULT] hap=all mut=Indels reg=all count=1065 len=126695273 perMb=8.406\n",
      "[RESULT] hap=maternal mut=all reg=hard count=348 len=11708185 perMb=29.723\n",
      "[RESULT] hap=maternal mut=Indels reg=easy count=0 len=21392665 perMb=0.000\n",
      "[RESULT] hap=maternal mut=SNVs reg=hard count=1 len=5249467 perMb=0.190\n",
      "[RESULT] hap=maternal mut=SNVs reg=all count=30 len=77297733 perMb=0.388\n",
      "[RESULT] hap=all mut=SNVs reg=hard count=16 len=24946862 perMb=0.641\n",
      "[RESULT] hap=maternal mut=all reg=easy count=18 len=87023261 perMb=0.207\n",
      "[RESULT] hap=paternal mut=all reg=easy count=3 len=31010908 perMb=0.097\n",
      "[RESULT] hap=maternal mut=Indels reg=all count=217 len=42603212 perMb=5.094\n",
      "[RESULT] hap=all mut=Indels reg=all count=1074 len=137502700 perMb=7.811\n",
      "[RESULT] hap=maternal mut=Indels reg=hard count=119 len=14945821 perMb=7.962\n",
      "[RESULT] hap=maternal mut=SNVs reg=hard count=5 len=7244238 perMb=0.690\n",
      "[RESULT] hap=paternal mut=all reg=all count=393 len=51333548 perMb=7.656\n",
      "[RESULT] hap=paternal mut=all reg=easy count=6 len=37041704 perMb=0.162\n",
      "[RESULT] hap=maternal mut=Indels reg=all count=225 len=46332509 perMb=4.856\n",
      "[RESULT] hap=all mut=Indels reg=easy count=5 len=128573090 perMb=0.039\n",
      "[RESULT] hap=all mut=SNVs reg=hard count=15 len=28076014 perMb=0.534\n",
      "[RESULT] hap=maternal mut=Indels reg=easy count=2 len=37353745 perMb=0.054\n",
      "[RESULT] hap=all mut=Indels reg=hard count=789 len=28382135 perMb=27.799\n",
      "[RESULT] hap=maternal mut=SNVs reg=easy count=17 len=65589548 perMb=0.259\n",
      "[RESULT] hap=all mut=Indels reg=all count=1167 len=156673292 perMb=7.449\n",
      "[RESULT] hap=all mut=all reg=hard count=1911 len=56875286 perMb=33.600\n",
      "[RESULT] hap=maternal mut=all reg=all count=533 len=130030136 perMb=4.099\n",
      "[RESULT] hap=paternal mut=all reg=hard count=638 len=23117874 perMb=27.598\n",
      "[RESULT] hap=all mut=Indels reg=easy count=8 len=127839355 perMb=0.063\n",
      "[RESULT] hap=all mut=Indels reg=hard count=913 len=33054933 perMb=27.621\n",
      "[RESULT] hap=all mut=SNVs reg=all count=58 len=227278701 perMb=0.255\n",
      "[RESULT] hap=maternal mut=Indels reg=easy count=0 len=39088271 perMb=0.000\n",
      "[RESULT] hap=paternal mut=SNVs reg=all count=13 len=54128782 perMb=0.240\n",
      "[RESULT] hap=maternal mut=all reg=all count=431 len=79421573 perMb=5.427\n",
      "[RESULT] hap=maternal mut=all reg=hard count=577 len=23251481 perMb=24.816\n",
      "[RESULT] hap=paternal mut=all reg=hard count=674 len=28168795 perMb=23.927\n",
      "[RESULT] hap=maternal mut=Indels reg=hard count=215 len=5249467 perMb=40.957\n",
      "[RESULT] hap=paternal mut=all reg=easy count=14 len=42146770 perMb=0.332\n",
      "[RESULT] hap=maternal mut=SNVs reg=all count=18 len=110274742 perMb=0.163\n",
      "[RESULT] hap=paternal mut=SNVs reg=all count=17 len=65210499 perMb=0.261\n",
      "[RESULT] hap=paternal mut=all reg=all count=496 len=54528274 perMb=9.096\n",
      "[RESULT] hap=maternal mut=SNVs reg=hard count=13 len=11708185 perMb=1.110\n",
      "[RESULT] hap=maternal mut=Indels reg=hard count=225 len=7244238 perMb=31.059\n",
      "[RESULT] hap=all mut=Indels reg=easy count=11 len=103592288 perMb=0.106\n",
      "[RESULT] hap=maternal mut=Indels reg=all count=337 len=77297733 perMb=4.360\n",
      "[RESULT] hap=paternal mut=all reg=all count=742 len=79288358 perMb=9.358\n",
      "[RESULT] hap=paternal mut=SNVs reg=easy count=1 len=31010908 perMb=0.032\n",
      "[RESULT] hap=all mut=SNVs reg=easy count=7 len=184514589 perMb=0.038\n",
      "[RESULT] hap=maternal mut=all reg=easy count=9 len=106595953 perMb=0.084\n",
      "[RESULT] hap=all mut=Indels reg=easy count=2 len=112555838 perMb=0.018\n",
      "[RESULT] hap=all mut=Indels reg=hard count=873 len=28062274 perMb=31.109\n",
      "[RESULT] hap=maternal mut=all reg=easy count=10 len=57815631 perMb=0.173\n",
      "[RESULT] hap=paternal mut=all reg=hard count=379 len=9186778 perMb=41.255\n",
      "[RESULT] hap=paternal mut=all reg=easy count=6 len=33266960 perMb=0.180[RESULT] hap=maternal mut=all reg=all count=621 len=121662463 perMb=5.104\n",
      "\n",
      "[RESULT] hap=paternal mut=SNVs reg=all count=13 len=51333548 perMb=0.253\n",
      "[RESULT] hap=maternal mut=Indels reg=easy count=2 len=65589548 perMb=0.030\n",
      "[RESULT] hap=paternal mut=SNVs reg=easy count=2 len=37041704 perMb=0.054\n",
      "[RESULT] hap=maternal mut=SNVs reg=easy count=11 len=87023261 perMb=0.126\n",
      "[RESULT] hap=all mut=Indels reg=hard count=971 len=32457771 perMb=29.916\n",
      "[RESULT] hap=paternal mut=all reg=easy count=9 len=43680063 perMb=0.206\n",
      "[RESULT] hap=paternal mut=SNVs reg=hard count=12 len=23117874 perMb=0.519\n",
      "[RESULT] hap=maternal mut=all reg=all count=507 len=100787742 perMb=5.030\n",
      "[RESULT] hap=all mut=Indels reg=easy count=6 len=128597278 perMb=0.047[RESULT] hap=paternal mut=Indels reg=all count=628 len=54128782 perMb=11.602\n",
      "\n",
      "[RESULT] hap=paternal mut=all reg=hard count=490 len=21261314 perMb=23.047\n",
      "[RESULT] hap=maternal mut=all reg=hard count=421 len=21605942 perMb=19.485\n",
      "[RESULT] hap=paternal mut=SNVs reg=all count=12 len=54528274 perMb=0.220\n",
      "[RESULT] hap=maternal mut=all reg=hard count=524 len=23434183 perMb=22.360\n",
      "[RESULT] hap=maternal mut=SNVs reg=all count=14 len=79421573 perMb=0.176\n",
      "[RESULT] hap=maternal mut=Indels reg=hard count=335 len=11708185 perMb=28.612\n",
      "[RESULT] hap=maternal mut=SNVs reg=all count=15 len=130030136 perMb=0.115\n",
      "[RESULT] hap=paternal mut=SNVs reg=easy count=11 len=42146770 perMb=0.261\n",
      "[RESULT] hap=paternal mut=SNVs reg=hard count=15 len=28168795 perMb=0.533\n",
      "[RESULT] hap=paternal mut=all reg=all count=366 len=39129207 perMb=9.354\n",
      "[RESULT] hap=all mut=Indels reg=hard count=1054 len=23102985 perMb=45.622\n",
      "[RESULT] hap=paternal mut=Indels reg=all count=663 len=65210499 perMb=10.167\n",
      "[RESULT] hap=paternal mut=all reg=hard count=733 len=35608295 perMb=20.585\n",
      "[RESULT] hap=maternal mut=all reg=easy count=7 len=98839394 perMb=0.071\n",
      "[RESULT] hap=maternal mut=all reg=all count=404 len=61950816 perMb=6.521\n",
      "[RESULT] hap=paternal mut=Indels reg=easy count=2 len=31010908 perMb=0.064\n",
      "[RESULT] hap=maternal mut=SNVs reg=hard count=7 len=23251481 perMb=0.301\n",
      "[RESULT] hap=paternal mut=SNVs reg=all count=15 len=79288358 perMb=0.189\n",
      "[RESULT] hap=paternal mut=SNVs reg=easy count=1 len=33266960 perMb=0.030\n",
      "[RESULT] hap=maternal mut=Indels reg=all count=577 len=110274742 perMb=5.232\n",
      "[RESULT] hap=maternal mut=all reg=easy count=12 len=82826237 perMb=0.145\n",
      "[RESULT] hap=all mut=Indels reg=hard count=1072 len=24946862 perMb=42.971\n",
      "[RESULT] hap=paternal mut=all reg=easy count=3 len=27322151 perMb=0.110\n",
      "[RESULT] hap=maternal mut=SNVs reg=easy count=7 len=57815631 perMb=0.121\n",
      "[RESULT] hap=maternal mut=all reg=all count=178 len=41709961 perMb=4.268\n",
      "[RESULT] hap=all mut=SNVs reg=easy count=25 len=170403415 perMb=0.147\n",
      "[RESULT] hap=paternal mut=SNVs reg=hard count=2 len=9186778 perMb=0.218\n",
      "[RESULT] hap=maternal mut=SNVs reg=easy count=4 len=106595953 perMb=0.038\n",
      "[RESULT] hap=paternal mut=Indels reg=easy count=4 len=37041704 perMb=0.108\n",
      "[RESULT] hap=paternal mut=Indels reg=all count=380 len=51333548 perMb=7.403\n",
      "[RESULT] hap=maternal mut=all reg=easy count=30 len=52951896 perMb=0.567\n",
      "[RESULT] hap=maternal mut=all reg=easy count=1 len=35979937 perMb=0.028\n",
      "[RESULT] hap=paternal mut=SNVs reg=hard count=11 len=21261314 perMb=0.517\n",
      "[RESULT] hap=paternal mut=Indels reg=hard count=626 len=23117874 perMb=27.079\n",
      "[RESULT] hap=paternal mut=SNVs reg=easy count=1 len=43680063 perMb=0.023\n",
      "[DONE] summary written to /tmp/mutrate_896r0tio/chr16/mutation_rates_summary.tsv\n",
      "[INFO] cleaning tmpdir: /tmp/mutrate_896r0tio/chr16\n",
      "[RESULT] hap=paternal mut=all reg=hard count=363 len=11807056 perMb=30.744\n",
      "[RESULT] hap=all mut=SNVs reg=hard count=15 len=44208114 perMb=0.339\n",
      "[RESULT] hap=paternal mut=Indels reg=all count=484 len=54528274 perMb=8.876\n",
      "[OK] chr16\n",
      "[CMD] padding+merge -> /tmp/mutrate_896r0tio/chr17/flagger_ont.padded.bed\n",
      "[RESULT] hap=paternal mut=SNVs reg=all count=8 len=39129207 perMb=0.204\n",
      "[CMD] padding+merge -> /tmp/mutrate_896r0tio/chr17/flagger_hifi.padded.bed\n",
      "[RESULT] hap=all mut=Indels reg=all count=1524 len=228722703 perMb=6.663\n",
      "[CMD] padding+merge -> /tmp/mutrate_896r0tio/chr17/problematic.padded.bed\n",
      "[CMD] combining excludes -> /tmp/mutrate_896r0tio/chr17/combined_exclude.bed\n",
      "[RESULT] hap=all mut=Indels reg=hard count=1162 len=28076014 perMb=41.388\n",
      "[RESULT] hap=maternal mut=all reg=hard count=495 len=17961505 perMb=27.559\n",
      "[RESULT] hap=maternal mut=all reg=hard count=614 len=22823069 perMb=26.903\n",
      "[RESULT] hap=maternal mut=SNVs reg=hard count=7 len=21605942 perMb=0.324\n",
      "[RESULT] hap=maternal mut=Indels reg=easy count=7 len=87023261 perMb=0.080\n",
      "[RESULT] hap=maternal mut=all reg=all count=348 len=69351345 perMb=5.018\n",
      "[RESULT] hap=maternal mut=all reg=hard count=177 len=5730024 perMb=30.890\n",
      "[RESULT] hap=maternal mut=SNVs reg=all count=11 len=100787742 perMb=0.109\n",
      "[RESULT] hap=maternal mut=Indels reg=all count=417 len=79421573 perMb=5.250[RESULT] hap=maternal mut=SNVs reg=all count=9 len=121662463 perMb=0.074\n",
      "\n",
      "[RESULT] hap=paternal mut=Indels reg=hard count=659 len=28168795 perMb=23.395\n",
      "[DONE] summary written to /tmp/mutrate_896r0tio/chr15/mutation_rates_summary.tsv\n",
      "[INFO] cleaning tmpdir: /tmp/mutrate_896r0tio/chr15\n",
      "[RESULT] hap=maternal mut=SNVs reg=all count=2 len=41709961 perMb=0.048\n",
      "[CMD] padding+merge -> /tmp/mutrate_896r0tio/chr18/flagger_ont.padded.bed\n",
      "[OK] chr15\n",
      "[RESULT] hap=paternal mut=SNVs reg=easy count=2 len=27322151 perMb=0.073\n",
      "[CMD] padding+merge -> /tmp/mutrate_896r0tio/chr18/flagger_hifi.padded.bed\n",
      "[RESULT] hap=maternal mut=all reg=hard count=374 len=8998920 perMb=41.561\n",
      "[RESULT] hap=paternal mut=Indels reg=easy count=3 len=42146770 perMb=0.071\n",
      "[CMD] padding+merge -> /tmp/mutrate_896r0tio/chr18/problematic.padded.bed\n",
      "[CMD] combining excludes -> /tmp/mutrate_896r0tio/chr18/combined_exclude.bed\n",
      "[RESULT] hap=paternal mut=Indels reg=easy count=5 len=33266960 perMb=0.150\n",
      "[RESULT] hap=paternal mut=SNVs reg=hard count=14 len=35608295 perMb=0.393\n",
      "[RESULT] hap=maternal mut=SNVs reg=all count=26 len=61950816 perMb=0.420\n",
      "[RESULT] hap=maternal mut=SNVs reg=hard count=11 len=23434183 perMb=0.469\n",
      "[RESULT] hap=paternal mut=Indels reg=all count=727 len=79288358 perMb=9.169\n",
      "[RESULT] hap=maternal mut=Indels reg=all count=518 len=130030136 perMb=3.984\n",
      "[RESULT] hap=maternal mut=SNVs reg=easy count=1 len=35979937 perMb=0.028\n",
      "[RESULT] hap=maternal mut=all reg=easy count=4 len=58924678 perMb=0.068\n",
      "[RESULT] hap=paternal mut=SNVs reg=hard count=6 len=11807056 perMb=0.508\n",
      "[RESULT] hap=maternal mut=Indels reg=easy count=3 len=57815631 perMb=0.052\n",
      "[RESULT] hap=paternal mut=Indels reg=all count=358 len=39129207 perMb=9.149\n",
      "[RESULT] hap=maternal mut=SNVs reg=easy count=7 len=82826237 perMb=0.085\n",
      "[RESULT] hap=maternal mut=Indels reg=hard count=570 len=23251481 perMb=24.515\n",
      "[RESULT] hap=paternal mut=Indels reg=hard count=479 len=21261314 perMb=22.529\n",
      "[DONE] summary written to /tmp/mutrate_896r0tio/chr14/mutation_rates_summary.tsv\n",
      "[INFO] cleaning tmpdir: /tmp/mutrate_896r0tio/chr14\n",
      "[RESULT] hap=maternal mut=SNVs reg=hard count=1 len=5730024 perMb=0.175\n",
      "[RESULT] hap=paternal mut=all reg=all count=214 len=20174983 perMb=10.607[RESULT] hap=paternal mut=Indels reg=hard count=377 len=9186778 perMb=41.037\n",
      "\n",
      "[DONE] summary written to /tmp/mutrate_896r0tio/chr13/mutation_rates_summary.tsv\n",
      "[INFO] cleaning tmpdir: /tmp/mutrate_896r0tio/chr13\n",
      "[CMD] padding+merge -> /tmp/mutrate_896r0tio/chr19/flagger_ont.padded.bed\n",
      "[RESULT] hap=maternal mut=SNVs reg=easy count=22 len=52951896 perMb=0.415\n",
      "[CMD] padding+merge -> [RESULT] hap=maternal mut=Indels reg=all count=176 len=41709961 perMb=4.220/tmp/mutrate_896r0tio/chr19/flagger_hifi.padded.bed\n",
      "\n",
      "[OK] chr14\n",
      "[OK] chr13\n",
      "[CMD] padding+merge -> /tmp/mutrate_896r0tio/chr19/problematic.padded.bed\n",
      "[CMD] combining excludes -> /tmp/mutrate_896r0tio/chr19/combined_exclude.bed\n",
      "[RESULT] hap=paternal mut=Indels reg=easy count=8 len=43680063 perMb=0.183[CMD] padding+merge ->\n",
      " /tmp/mutrate_896r0tio/chr20/flagger_ont.padded.bed\n",
      "[CMD] padding+merge -> /tmp/mutrate_896r0tio/chr20/flagger_hifi.padded.bed\n",
      "[RESULT] hap=maternal mut=SNVs reg=easy count=3 len=98839394 perMb=0.030\n",
      "[CMD] padding+merge -> /tmp/mutrate_896r0tio/chr20/problematic.padded.bed\n",
      "[CMD] combining excludes -> /tmp/mutrate_896r0tio/chr20/combined_exclude.bed\n",
      "[RESULT] hap=paternal mut=all reg=easy count=4 len=16571302 perMb=0.241\n",
      "[RESULT] hap=maternal mut=all reg=hard count=345 len=10426667 perMb=33.088\n",
      "[RESULT] hap=paternal mut=Indels reg=easy count=1 len=27322151 perMb=0.037\n",
      "[RESULT] hap=maternal mut=SNVs reg=all count=8 len=69351345 perMb=0.115\n",
      "[RESULT] hap=maternal mut=Indels reg=easy count=5 len=106595953 perMb=0.047\n",
      "[RESULT] hap=maternal mut=Indels reg=hard count=414 len=21605942 perMb=19.161\n",
      "[RESULT] hap=paternal mut=all reg=hard count=210 len=3603681 perMb=58.274\n",
      "[RESULT] hap=maternal mut=Indels reg=easy count=0 len=35979937 perMb=0.000\n",
      "[RESULT] hap=paternal mut=all reg=all count=518 len=59071261 perMb=8.769\n",
      "[RESULT] hap=paternal mut=SNVs reg=all count=0 len=20174983 perMb=0.000\n",
      "[RESULT] hap=all mut=SNVs reg=hard count=33 len=56875286 perMb=0.580\n",
      "[RESULT] hap=maternal mut=SNVs reg=hard count=4 len=17961505 perMb=0.223\n",
      "[RESULT] hap=maternal mut=SNVs reg=hard count=4 len=8998920 perMb=0.444\n",
      "[RESULT] hap=paternal mut=Indels reg=hard count=719 len=35608295 perMb=20.192\n",
      "[DONE] summary written to /tmp/mutrate_896r0tio/chr9/mutation_rates_summary.tsv\n",
      "[INFO] cleaning tmpdir: /tmp/mutrate_896r0tio/chr9\n",
      "[RESULT] hap=paternal mut=Indels reg=hard count=357 len=11807056 perMb=30.236\n",
      "[RESULT] hap=maternal mut=Indels reg=all count=496 len=100787742 perMb=4.921[RESULT] hap=paternal mut=SNVs reg=easy count=0 len=16571302 perMb=0.000\n",
      "\n",
      "[RESULT] hap=all mut=Indels reg=all count=1887 len=227278701 perMb=8.303\n",
      "[DONE] summary written to /tmp/mutrate_896r0tio/chr11/mutation_rates_summary.tsv\n",
      "[INFO] cleaning tmpdir: /tmp/mutrate_896r0tio/chr11\n",
      "[RESULT] hap=all mut=Indels reg=easy count=11 len=184514589 perMb=0.060[RESULT] hap=maternal mut=Indels reg=all count=378 len=61950816 perMb=6.102[RESULT] hap=maternal mut=Indels reg=hard count=176 len=5730024 perMb=30.715\n",
      "\n",
      "\n",
      "[CMD] padding+merge -> /tmp/mutrate_896r0tio/chr21/flagger_ont.padded.bed\n",
      "[CMD] padding+merge -> /tmp/mutrate_896r0tio/chr21/flagger_hifi.padded.bed\n",
      "[CMD] padding+merge -> /tmp/mutrate_896r0tio/chr22/flagger_ont.padded.bed\n",
      "[CMD] padding+merge -> /tmp/mutrate_896r0tio/chr21/problematic.padded.bed\n",
      "[CMD] padding+merge -> /tmp/mutrate_896r0tio/chr22/flagger_hifi.padded.bed\n",
      "[OK] chr9\n",
      "[OK] chr11\n",
      "[RESULT] hap=paternal mut=all reg=all count=915 len=95792739 perMb=9.552[CMD] padding+merge -> \n",
      "/tmp/mutrate_896r0tio/chr22/problematic.padded.bed\n",
      "[CMD] combining excludes -> /tmp/mutrate_896r0tio/chr21/combined_exclude.bed\n",
      "[RESULT] hap=maternal mut=SNVs reg=easy count=1 len=58924678 perMb=0.017\n",
      "[CMD] combining excludes -> /tmp/mutrate_896r0tio/chr22/combined_exclude.bed\n",
      "[RESULT] hap=paternal mut=SNVs reg=hard count=0 len=3603681 perMb=0.000\n",
      "[RESULT] hap=paternal mut=Indels reg=all count=214 len=20174983 perMb=10.607\n",
      "[RESULT] hap=maternal mut=SNVs reg=hard count=6 len=22823069 perMb=0.263\n",
      "[RESULT] hap=paternal mut=all reg=easy count=6 len=47622270 perMb=0.126\n",
      "[RESULT] hap=maternal mut=Indels reg=hard count=513 len=23434183 perMb=21.891\n",
      "[RESULT] hap=maternal mut=Indels reg=all count=612 len=121662463 perMb=5.030\n",
      "[RESULT] hap=paternal mut=Indels reg=easy count=4 len=16571302 perMb=0.241\n",
      "[RESULT] hap=paternal mut=all reg=all count=294 len=24860092 perMb=11.826\n",
      "[RESULT] hap=paternal mut=Indels reg=hard count=210 len=3603681 perMb=58.274\n",
      "[RESULT] hap=maternal mut=Indels reg=easy count=8 len=52951896 perMb=0.151\n",
      "[DONE] summary written to /tmp/mutrate_896r0tio/chr10/mutation_rates_summary.tsv\n",
      "[INFO] cleaning tmpdir: /tmp/mutrate_896r0tio/chr10\n",
      "[RESULT] hap=paternal mut=all reg=easy count=4 len=19912140 perMb=0.201\n",
      "[RESULT] hap=maternal mut=SNVs reg=hard count=7 len=10426667 perMb=0.671[RESULT] hap=maternal mut=Indels reg=easy count=5 len=82826237 perMb=0.060\n",
      "\n",
      "[CMD] padding+merge -> /tmp/mutrate_896r0tio/chrX/flagger_ont.padded.bed\n",
      "[CMD] padding+merge -> /tmp/mutrate_896r0tio/chrX/flagger_hifi.padded.bed\n",
      "[CMD] padding+merge -> /tmp/mutrate_896r0tio/chrX/problematic.padded.bed\n",
      "[CMD] combining excludes -> /tmp/mutrate_896r0tio/chrX/combined_exclude.bed\n",
      "[RESULT] hap=maternal mut=Indels reg=all count=340 len=69351345 perMb=4.903\n",
      "[OK] chr10\n",
      "[RESULT] hap=paternal mut=all reg=hard count=512 len=11448991 perMb=44.720\n",
      "[RESULT] hap=paternal mut=all reg=hard count=290 len=4947952 perMb=58.610\n",
      "[RESULT] hap=paternal mut=SNVs reg=all count=14 len=59071261 perMb=0.237\n",
      "[RESULT] hap=paternal mut=SNVs reg=all count=15 len=24860092 perMb=0.603\n",
      "[RESULT] hap=paternal mut=all reg=easy count=4 len=76575901 perMb=0.052\n",
      "[RESULT] hap=maternal mut=Indels reg=hard count=370 len=8998920 perMb=41.116\n",
      "[RESULT] hap=paternal mut=all reg=all count=699 len=64744457 perMb=10.796\n",
      "[RESULT] hap=paternal mut=SNVs reg=easy count=1 len=19912140 perMb=0.050\n",
      "[RESULT] hap=maternal mut=Indels reg=easy count=4 len=98839394 perMb=0.040\n",
      "[RESULT] hap=maternal mut=Indels reg=easy count=3 len=58924678 perMb=0.051\n",
      "[RESULT] hap=maternal mut=Indels reg=hard count=491 len=17961505 perMb=27.336\n",
      "[RESULT] hap=paternal mut=all reg=all count=495 len=59509384 perMb=8.318\n",
      "[RESULT] hap=paternal mut=SNVs reg=hard count=14 len=4947952 perMb=2.829\n",
      "[RESULT] hap=paternal mut=SNVs reg=easy count=1 len=47622270 perMb=0.021\n",
      "[RESULT] hap=paternal mut=Indels reg=all count=279 len=24860092 perMb=11.223\n",
      "[RESULT] hap=all mut=Indels reg=hard count=1513 len=44208114 perMb=34.224\n",
      "[RESULT] hap=paternal mut=Indels reg=easy count=3 len=19912140 perMb=0.151\n",
      "[RESULT] hap=maternal mut=Indels reg=hard count=338 len=10426667 perMb=32.417\n",
      "[RESULT] hap=maternal mut=all reg=all count=697 len=133422841 perMb=5.224\n",
      "[RESULT] hap=paternal mut=all reg=all count=835 len=87321947 perMb=9.562\n",
      "[RESULT] hap=paternal mut=all reg=easy count=10 len=45013118 perMb=0.222\n",
      "[RESULT] hap=paternal mut=Indels reg=hard count=276 len=4947952 perMb=55.781\n",
      "[RESULT] hap=paternal mut=SNVs reg=hard count=13 len=11448991 perMb=1.135\n",
      "[DONE] summary written to /tmp/mutrate_896r0tio/chr4/mutation_rates_summary.tsv\n",
      "[INFO] cleaning tmpdir: /tmp/mutrate_896r0tio/chr4\n",
      "[RESULT] hap=all mut=Indels reg=easy count=11 len=170403415 perMb=0.065\n",
      "[RESULT] hap=paternal mut=all reg=easy count=5 len=50640392 perMb=0.099\n",
      "[RESULT] hap=paternal mut=all reg=hard count=911 len=19216838 perMb=47.406\n",
      "[RESULT] hap=paternal mut=Indels reg=all count=504 len=59071261 perMb=8.532\n",
      "[RESULT] hap=paternal mut=SNVs reg=all count=17 len=95792739 perMb=0.177\n",
      "[OK] chr4\n",
      "[RESULT] hap=maternal mut=Indels reg=hard count=608 len=22823069 perMb=26.640\n",
      "[RESULT] hap=paternal mut=all reg=all count=273 len=34972901 perMb=7.806\n",
      "[RESULT] hap=paternal mut=all reg=hard count=485 len=14496266 perMb=33.457\n",
      "[RESULT] hap=paternal mut=all reg=easy count=3 len=29733696 perMb=0.101\n",
      "[RESULT] hap=paternal mut=SNVs reg=all count=12 len=59509384 perMb=0.202\n",
      "[RESULT] hap=paternal mut=Indels reg=easy count=5 len=47622270 perMb=0.105\n",
      "[RESULT] hap=paternal mut=all reg=easy count=3 len=69672600 perMb=0.043\n",
      "[RESULT] hap=paternal mut=all reg=hard count=270 len=5239205 perMb=51.535\n",
      "[RESULT] hap=paternal mut=all reg=hard count=694 len=14104065 perMb=49.206\n",
      "[RESULT] hap=maternal mut=all reg=easy count=13 len=114885047 perMb=0.113\n",
      "[RESULT] hap=paternal mut=SNVs reg=all count=7 len=34972901 perMb=0.200\n",
      "[RESULT] hap=paternal mut=SNVs reg=all count=12 len=64744457 perMb=0.185\n",
      "[RESULT] hap=paternal mut=SNVs reg=easy count=7 len=45013118 perMb=0.156\n",
      "[RESULT] hap=paternal mut=SNVs reg=easy count=2 len=76575901 perMb=0.026\n",
      "[RESULT] hap=paternal mut=Indels reg=hard count=499 len=11448991 perMb=43.585\n",
      "[DONE] summary written to /tmp/mutrate_896r0tio/chr7/mutation_rates_summary.tsv\n",
      "[INFO] cleaning tmpdir: /tmp/mutrate_896r0tio/chr7\n",
      "[RESULT] hap=paternal mut=SNVs reg=easy count=2 len=29733696 perMb=0.067\n",
      "[OK] chr7\n",
      "[RESULT] hap=paternal mut=SNVs reg=hard count=5 len=5239205 perMb=0.954\n",
      "[RESULT] hap=paternal mut=Indels reg=all count=266 len=34972901 perMb=7.606\n",
      "[RESULT] hap=paternal mut=SNVs reg=hard count=5 len=14496266 perMb=0.345\n",
      "[RESULT] hap=paternal mut=Indels reg=all count=483 len=59509384 perMb=8.116\n",
      "[RESULT] hap=paternal mut=SNVs reg=easy count=2 len=50640392 perMb=0.039\n",
      "[INFO] Running counts (this may take a while depending on VCF size).\n",
      "[RESULT] hap=paternal mut=all reg=hard count=832 len=17649347 perMb=47.141\n",
      "[RESULT] hap=paternal mut=Indels reg=easy count=1 len=29733696 perMb=0.034\n",
      "[RESULT] hap=maternal mut=all reg=hard count=684 len=18537794 perMb=36.898\n",
      "[RESULT] hap=all mut=all reg=all count=899 len=68588485 perMb=13.107\n",
      "[INFO] Running counts (this may take a while depending on VCF size).\n",
      "[RESULT] hap=paternal mut=SNVs reg=all count=8 len=87321947 perMb=0.092\n",
      "[RESULT] hap=maternal mut=SNVs reg=all count=13 len=133422841 perMb=0.097\n",
      "[RESULT] hap=all mut=all reg=all count=394 len=66350853 perMb=5.938\n",
      "[RESULT] hap=paternal mut=Indels reg=hard count=265 len=5239205 perMb=50.580\n",
      "[DONE] summary written to /tmp/mutrate_896r0tio/chr6/mutation_rates_summary.tsv\n",
      "[INFO] cleaning tmpdir: /tmp/mutrate_896r0tio/chr6\n",
      "[RESULT] hap=all mut=Indels reg=hard count=1878 len=56875286 perMb=33.020\n",
      "[RESULT] hap=paternal mut=SNVs reg=hard count=15 len=19216838 perMb=0.781\n",
      "[RESULT] hap=paternal mut=Indels reg=easy count=3 len=45013118 perMb=0.067\n",
      "[RESULT] hap=maternal mut=all reg=all count=482 len=87818754 perMb=5.489\n",
      "[RESULT] hap=paternal mut=Indels reg=all count=898 len=95792739 perMb=9.374\n",
      "[INFO] Running counts (this may take a while depending on VCF size).\n",
      "[OK] chr6\n",
      "[RESULT] hap=all mut=all reg=easy count=51 len=50550957 perMb=1.009\n",
      "[RESULT] hap=all mut=all reg=all count=691 len=51871506 perMb=13.321\n",
      "[INFO] Running counts (this may take a while depending on VCF size).\n",
      "[RESULT] hap=paternal mut=SNVs reg=hard count=10 len=14104065 perMb=0.709\n",
      "[INFO] Running counts (this may take a while depending on VCF size).\n",
      "[RESULT] hap=all mut=all reg=all count=403 len=63397920 perMb=6.357\n",
      "[RESULT] hap=paternal mut=Indels reg=all count=687 len=64744457 perMb=10.611\n",
      "[RESULT] hap=all mut=all reg=easy count=9 len=47999872 perMb=0.188\n",
      "[RESULT] hap=all mut=all reg=all count=278 len=39285226 perMb=7.076\n",
      "[RESULT] hap=paternal mut=Indels reg=hard count=480 len=14496266 perMb=33.112\n",
      "[DONE] summary written to /tmp/mutrate_896r0tio/chr5/mutation_rates_summary.tsv\n",
      "[INFO] cleaning tmpdir: /tmp/mutrate_896r0tio/chr5\n",
      "[RESULT] hap=paternal mut=SNVs reg=easy count=0 len=69672600 perMb=0.000\n",
      "[INFO] Running counts (this may take a while depending on VCF size).\n",
      "[RESULT] hap=maternal mut=all reg=easy count=27 len=54966983 perMb=0.491\n",
      "[RESULT] hap=all mut=all reg=all count=173 len=35653800 perMb=4.852\n",
      "[RESULT] hap=all mut=all reg=hard count=343 len=15799896 perMb=21.709\n",
      "[RESULT] hap=all mut=all reg=easy count=4 len=22526431 perMb=0.178\n",
      "[RESULT] hap=maternal mut=SNVs reg=easy count=6 len=114885047 perMb=0.052\n",
      "[OK] chr5\n",
      "[RESULT] hap=all mut=SNVs reg=all count=57 len=66350853 perMb=0.859\n",
      "[RESULT] hap=all mut=all reg=easy count=1 len=48789973 perMb=0.020\n",
      "[RESULT] hap=all mut=all reg=easy count=4 len=32103137 perMb=0.125\n",
      "[RESULT] hap=all mut=all reg=easy count=11 len=21939439 perMb=0.501\n",
      "[RESULT] hap=all mut=all reg=hard count=274 len=16758795 perMb=16.350\n",
      "[RESULT] hap=paternal mut=Indels reg=easy count=2 len=76575901 perMb=0.026\n",
      "[RESULT] hap=all mut=SNVs reg=all count=3 len=39285226 perMb=0.076\n",
      "[RESULT] hap=paternal mut=Indels reg=easy count=3 len=50640392 perMb=0.059\n",
      "[RESULT] hap=all mut=all reg=hard count=162 len=13714361 perMb=11.812\n",
      "[RESULT] hap=all mut=all reg=hard count=890 len=20588613 perMb=43.228\n",
      "[RESULT] hap=all mut=SNVs reg=all count=29 len=35653800 perMb=0.813\n",
      "[RESULT] hap=all mut=SNVs reg=easy count=46 len=50550957 perMb=0.910\n",
      "[RESULT] hap=all mut=SNVs reg=all count=15 len=68588485 perMb=0.219\n",
      "[RESULT] hap=all mut=SNVs reg=easy count=2 len=22526431 perMb=0.089\n",
      "[RESULT] hap=all mut=all reg=hard count=402 len=14607947 perMb=27.519\n",
      "[RESULT] hap=all mut=SNVs reg=easy count=9 len=21939439 perMb=0.410\n",
      "[RESULT] hap=maternal mut=all reg=hard count=455 len=32851771 perMb=13.850\n",
      "[RESULT] hap=all mut=SNVs reg=all count=14 len=63397920 perMb=0.221\n",
      "[RESULT] hap=maternal mut=SNVs reg=all count=38 len=87818754 perMb=0.433\n",
      "[RESULT] hap=paternal mut=SNVs reg=hard count=8 len=17649347 perMb=0.453\n",
      "[RESULT] hap=all mut=SNVs reg=hard count=1 len=16758795 perMb=0.060\n",
      "[RESULT] hap=all mut=all reg=hard count=687 len=19768369 perMb=34.752\n",
      "[RESULT] hap=all mut=SNVs reg=hard count=20 len=13714361 perMb=1.458\n",
      "[RESULT] hap=maternal mut=SNVs reg=hard count=7 len=18537794 perMb=0.378\n",
      "[RESULT] hap=all mut=Indels reg=all count=275 len=39285226 perMb=7.000\n",
      "[RESULT] hap=paternal mut=Indels reg=all count=827 len=87321947 perMb=9.471\n",
      "[RESULT] hap=all mut=Indels reg=all count=144 len=35653800 perMb=4.039\n",
      "[RESULT] hap=all mut=SNVs reg=all count=14 len=51871506 perMb=0.270\n",
      "[RESULT] hap=all mut=SNVs reg=hard count=11 len=15799896 perMb=0.696\n",
      "[RESULT] hap=maternal mut=Indels reg=all count=684 len=133422841 perMb=5.127\n",
      "[RESULT] hap=all mut=Indels reg=all count=337 len=66350853 perMb=5.079\n",
      "[RESULT] hap=paternal mut=Indels reg=hard count=684 len=14104065 perMb=48.497\n",
      "[RESULT] hap=all mut=Indels reg=easy count=2 len=21939439 perMb=0.091\n",
      "[DONE] summary written to /tmp/mutrate_896r0tio/chr12/mutation_rates_summary.tsv\n",
      "[INFO] cleaning tmpdir: /tmp/mutrate_896r0tio/chr12\n",
      "[RESULT] hap=all mut=Indels reg=easy count=2 len=22526431 perMb=0.089\n",
      "[RESULT] hap=all mut=SNVs reg=easy count=1 len=48789973 perMb=0.020\n",
      "[RESULT] hap=paternal mut=Indels reg=hard count=896 len=19216838 perMb=46.626\n",
      "[DONE] summary written to /tmp/mutrate_896r0tio/chr8/mutation_rates_summary.tsv\n",
      "[INFO] cleaning tmpdir: /tmp/mutrate_896r0tio/chr8\n",
      "[RESULT] hap=all mut=SNVs reg=easy count=4 len=47999872 perMb=0.083\n",
      "[OK] chr12\n",
      "[OK] chr8\n",
      "[RESULT] hap=all mut=Indels reg=hard count=142 len=13714361 perMb=10.354\n",
      "[RESULT] hap=all mut=Indels reg=hard count=273 len=16758795 perMb=16.290\n",
      "[RESULT] hap=maternal mut=all reg=all count=109 len=21880676 perMb=4.982\n",
      "[RESULT] hap=maternal mut=SNVs reg=easy count=23 len=54966983 perMb=0.418\n",
      "[RESULT] hap=maternal mut=all reg=all count=220 len=24989417 perMb=8.804\n",
      "[RESULT] hap=all mut=Indels reg=easy count=5 len=50550957 perMb=0.099\n",
      "[RESULT] hap=maternal mut=all reg=easy count=11 len=18404076 perMb=0.598\n",
      "[RESULT] hap=all mut=SNVs reg=easy count=1 len=32103137 perMb=0.031\n",
      "[RESULT] hap=all mut=SNVs reg=hard count=13 len=14607947 perMb=0.890\n",
      "[RESULT] hap=maternal mut=all reg=easy count=4 len=20455811 perMb=0.196\n",
      "[RESULT] hap=maternal mut=all reg=hard count=98 len=3476600 perMb=28.188\n",
      "[RESULT] hap=all mut=Indels reg=all count=389 len=63397920 perMb=6.136\n",
      "[RESULT] hap=paternal mut=Indels reg=easy count=3 len=69672600 perMb=0.043\n",
      "[RESULT] hap=maternal mut=SNVs reg=all count=11 len=21880676 perMb=0.503\n",
      "[RESULT] hap=maternal mut=all reg=hard count=216 len=4533606 perMb=47.644\n",
      "[RESULT] hap=maternal mut=Indels reg=easy count=7 len=114885047 perMb=0.061\n",
      "[RESULT] hap=maternal mut=SNVs reg=easy count=9 len=18404076 perMb=0.489\n",
      "[RESULT] hap=maternal mut=SNVs reg=all count=3 len=24989417 perMb=0.120\n",
      "[RESULT] hap=all mut=Indels reg=hard count=332 len=15799896 perMb=21.013\n",
      "[RESULT] hap=maternal mut=SNVs reg=hard count=15 len=32851771 perMb=0.457\n",
      "[RESULT] hap=all mut=SNVs reg=hard count=11 len=20588613 perMb=0.534\n",
      "[RESULT] hap=maternal mut=all reg=all count=247 len=51850423 perMb=4.764\n",
      "[RESULT] hap=maternal mut=Indels reg=all count=444 len=87818754 perMb=5.056\n",
      "[RESULT] hap=maternal mut=SNVs reg=hard count=2 len=3476600 perMb=0.575\n",
      "[RESULT] hap=all mut=Indels reg=all count=884 len=68588485 perMb=12.888\n",
      "[RESULT] hap=maternal mut=SNVs reg=easy count=2 len=20455811 perMb=0.098\n",
      "[RESULT] hap=all mut=Indels reg=easy count=0 len=48789973 perMb=0.000\n",
      "[RESULT] hap=maternal mut=Indels reg=all count=98 len=21880676 perMb=4.479\n",
      "[RESULT] hap=all mut=SNVs reg=hard count=13 len=19768369 perMb=0.658\n",
      "[RESULT] hap=all mut=Indels reg=all count=677 len=51871506 perMb=13.051\n",
      "[INFO] Running counts (this may take a while depending on VCF size).\n",
      "[RESULT] hap=maternal mut=Indels reg=easy count=2 len=18404076 perMb=0.109[RESULT] hap=maternal mut=SNVs reg=hard count=1 len=4533606 perMb=0.221\n",
      "\n",
      "[RESULT] hap=maternal mut=all reg=easy count=31 len=38469502 perMb=0.806\n",
      "[RESULT] hap=maternal mut=Indels reg=all count=217 len=24989417 perMb=8.684\n",
      "[RESULT] hap=all mut=all reg=all count=774 len=141738536 perMb=5.461\n",
      "[RESULT] hap=maternal mut=Indels reg=hard count=96 len=3476600 perMb=27.613\n",
      "[RESULT] hap=paternal mut=Indels reg=hard count=824 len=17649347 perMb=46.687\n",
      "[DONE] summary written to /tmp/mutrate_896r0tio/chr3/mutation_rates_summary.tsv\n",
      "[INFO] cleaning tmpdir: /tmp/mutrate_896r0tio/chr3\n",
      "[RESULT] hap=paternal mut=all reg=all count=64 len=13773124 perMb=4.647\n",
      "[RESULT] hap=maternal mut=Indels reg=easy count=2 len=20455811 perMb=0.098\n",
      "[RESULT] hap=maternal mut=Indels reg=easy count=4 len=54966983 perMb=0.073\n",
      "[RESULT] hap=all mut=Indels reg=hard count=389 len=14607947 perMb=26.629\n",
      "[RESULT] hap=paternal mut=all reg=easy count=0 len=3535363 perMb=0.000\n",
      "[RESULT] hap=maternal mut=all reg=hard count=216 len=13380921 perMb=16.142\n",
      "[OK] chr3\n",
      "[RESULT] hap=maternal mut=all reg=all count=289 len=50559899 perMb=5.716\n",
      "[RESULT] hap=paternal mut=all reg=hard count=64 len=10237761 perMb=6.251\n",
      "[RESULT] hap=maternal mut=SNVs reg=all count=38 len=51850423 perMb=0.733\n",
      "[RESULT] hap=maternal mut=Indels reg=hard count=215 len=4533606 perMb=47.424\n",
      "[RESULT] hap=all mut=Indels reg=easy count=5 len=47999872 perMb=0.104\n",
      "[RESULT] hap=maternal mut=Indels reg=hard count=677 len=18537794 perMb=36.520\n",
      "[RESULT] hap=paternal mut=SNVs reg=all count=18 len=13773124 perMb=1.307\n",
      "[RESULT] hap=paternal mut=all reg=all count=58 len=14295809 perMb=4.057\n",
      "[RESULT] hap=paternal mut=all reg=all count=849 len=95299862 perMb=8.909\n",
      "[RESULT] hap=all mut=Indels reg=easy count=3 len=32103137 perMb=0.093\n",
      "[RESULT] hap=paternal mut=SNVs reg=easy count=0 len=3535363 perMb=0.000\n",
      "[RESULT] hap=paternal mut=all reg=easy count=0 len=2070620 perMb=0.000\n",
      "[RESULT] hap=paternal mut=SNVs reg=hard count=18 len=10237761 perMb=1.758\n",
      "[RESULT] hap=paternal mut=all reg=hard count=58 len=12225189 perMb=4.744\n",
      "[RESULT] hap=maternal mut=SNVs reg=easy count=29 len=38469502 perMb=0.754\n",
      "[RESULT] hap=maternal mut=all reg=easy count=1 len=37374132 perMb=0.027\n",
      "[RESULT] hap=paternal mut=Indels reg=all count=46 len=13773124 perMb=3.340\n",
      "[RESULT] hap=paternal mut=SNVs reg=all count=0 len=14295809 perMb=0.000\n",
      "[RESULT] hap=all mut=all reg=easy count=23 len=109064061 perMb=0.211\n",
      "[RESULT] hap=maternal mut=Indels reg=hard count=440 len=32851771 perMb=13.393\n",
      "[RESULT] hap=paternal mut=Indels reg=easy count=0 len=3535363 perMb=0.000\n",
      "[RESULT] hap=paternal mut=SNVs reg=easy count=0 len=2070620 perMb=0.000\n",
      "[RESULT] hap=paternal mut=all reg=all count=1463 len=139459947 perMb=10.490\n",
      "[RESULT] hap=paternal mut=Indels reg=hard count=46 len=10237761 perMb=4.493[RESULT] hap=paternal mut=SNVs reg=hard count=0 len=12225189 perMb=0.000\n",
      "\n",
      "[DONE] summary written to /tmp/mutrate_896r0tio/chr21/mutation_rates_summary.tsv\n",
      "[INFO] cleaning tmpdir: /tmp/mutrate_896r0tio/chr21\n",
      "[RESULT] hap=maternal mut=SNVs reg=hard count=9 len=13380921 perMb=0.673\n",
      "[RESULT] hap=paternal mut=Indels reg=all count=58 len=14295809 perMb=4.057\n",
      "[RESULT] hap=maternal mut=all reg=hard count=288 len=13185767 perMb=21.842\n",
      "[RESULT] hap=all mut=Indels reg=hard count=879 len=20588613 perMb=42.694\n",
      "[RESULT] hap=maternal mut=Indels reg=all count=209 len=51850423 perMb=4.031\n",
      "[RESULT] hap=paternal mut=Indels reg=easy count=0 len=2070620 perMb=0.000\n",
      "[OK] chr21\n",
      "[RESULT] hap=maternal mut=SNVs reg=all count=14 len=50559899 perMb=0.277\n",
      "[RESULT] hap=maternal mut=all reg=all count=290 len=41291170 perMb=7.023\n",
      "[RESULT] hap=paternal mut=Indels reg=hard count=58 len=12225189 perMb=4.744\n",
      "[DONE] summary written to /tmp/mutrate_896r0tio/chr22/mutation_rates_summary.tsv\n",
      "[INFO] cleaning tmpdir: /tmp/mutrate_896r0tio/chr22\n",
      "[RESULT] hap=paternal mut=all reg=easy count=5 len=69629542 perMb=0.072\n",
      "[RESULT] hap=all mut=Indels reg=hard count=674 len=19768369 perMb=34.095\n",
      "[OK] chr22\n",
      "[RESULT] hap=maternal mut=Indels reg=easy count=2 len=38469502 perMb=0.052\n",
      "[RESULT] hap=maternal mut=all reg=all count=352 len=25612539 perMb=13.743\n",
      "[RESULT] hap=all mut=all reg=hard count=751 len=32674475 perMb=22.984[RESULT] hap=maternal mut=SNVs reg=easy count=1 len=37374132 perMb=0.027\n",
      "\n",
      "[RESULT] hap=maternal mut=all reg=easy count=9 len=27657843 perMb=0.325\n",
      "[RESULT] hap=all mut=SNVs reg=all count=38 len=141738536 perMb=0.268\n",
      "[RESULT] hap=maternal mut=Indels reg=hard count=207 len=13380921 perMb=15.470\n",
      "[RESULT] hap=paternal mut=all reg=all count=147 len=14500430 perMb=10.138\n",
      "[RESULT] hap=maternal mut=all reg=hard count=281 len=13633327 perMb=20.611\n",
      "[RESULT] hap=maternal mut=SNVs reg=hard count=13 len=13185767 perMb=0.986\n",
      "[RESULT] hap=maternal mut=all reg=easy count=2 len=18817422 perMb=0.106\n",
      "[RESULT] hap=maternal mut=SNVs reg=all count=8 len=41291170 perMb=0.194\n",
      "[RESULT] hap=maternal mut=Indels reg=all count=275 len=50559899 perMb=5.439\n",
      "[RESULT] hap=paternal mut=all reg=hard count=844 len=25670320 perMb=32.878\n",
      "[RESULT] hap=paternal mut=all reg=easy count=20 len=12081455 perMb=1.655\n",
      "[RESULT] hap=paternal mut=SNVs reg=all count=9 len=95299862 perMb=0.094\n",
      "[RESULT] hap=paternal mut=all reg=hard count=127 len=2418975 perMb=52.502\n",
      "[RESULT] hap=paternal mut=all reg=easy count=9 len=115436432 perMb=0.078\n",
      "[RESULT] hap=maternal mut=SNVs reg=easy count=4 len=27657843 perMb=0.145\n",
      "[RESULT] hap=paternal mut=SNVs reg=all count=19 len=14500430 perMb=1.310\n",
      "[RESULT] hap=maternal mut=Indels reg=easy count=0 len=37374132 perMb=0.000\n",
      "[RESULT] hap=maternal mut=all reg=hard count=350 len=6795117 perMb=51.508\n",
      "[RESULT] hap=all mut=SNVs reg=easy count=14 len=109064061 perMb=0.128\n",
      "[RESULT] hap=maternal mut=SNVs reg=all count=7 len=25612539 perMb=0.273\n",
      "[RESULT] hap=paternal mut=SNVs reg=easy count=17 len=12081455 perMb=1.407\n",
      "[RESULT] hap=maternal mut=SNVs reg=hard count=4 len=13633327 perMb=0.293\n",
      "[RESULT] hap=maternal mut=Indels reg=hard count=275 len=13185767 perMb=20.856\n",
      "[RESULT] hap=paternal mut=SNVs reg=hard count=2 len=2418975 perMb=0.827\n",
      "[RESULT] hap=maternal mut=Indels reg=all count=282 len=41291170 perMb=6.830\n",
      "[RESULT] hap=paternal mut=all reg=all count=114 len=12838021 perMb=8.880[RESULT] hap=paternal mut=Indels reg=all count=128 len=14500430 perMb=8.827\n",
      "\n",
      "[RESULT] hap=paternal mut=SNVs reg=easy count=1 len=69629542 perMb=0.014\n",
      "[RESULT] hap=maternal mut=SNVs reg=easy count=0 len=18817422 perMb=0.000\n",
      "[RESULT] hap=paternal mut=Indels reg=easy count=3 len=12081455 perMb=0.248\n",
      "[RESULT] hap=paternal mut=all reg=easy count=0 len=11415841 perMb=0.000\n",
      "[RESULT] hap=maternal mut=Indels reg=easy count=5 len=27657843 perMb=0.181\n",
      "[RESULT] hap=all mut=SNVs reg=hard count=24 len=32674475 perMb=0.735\n",
      "[RESULT] hap=paternal mut=Indels reg=hard count=125 len=2418975 perMb=51.675\n",
      "[DONE] summary written to /tmp/mutrate_896r0tio/chr18/mutation_rates_summary.tsv\n",
      "[INFO] cleaning tmpdir: /tmp/mutrate_896r0tio/chr18[RESULT] hap=all mut=Indels reg=all count=736 len=141738536 perMb=5.193\n",
      "\n",
      "[RESULT] hap=paternal mut=all reg=hard count=114 len=1422180 perMb=80.159\n",
      "[RESULT] hap=paternal mut=SNVs reg=all count=0 len=12838021 perMb=0.000\n",
      "[RESULT] hap=maternal mut=SNVs reg=hard count=7 len=6795117 perMb=1.030\n",
      "[RESULT] hap=maternal mut=Indels reg=all count=345 len=25612539 perMb=13.470\n",
      "[RESULT] hap=maternal mut=Indels reg=hard count=277 len=13633327 perMb=20.318\n",
      "[OK] chr18\n",
      "[RESULT] hap=paternal mut=SNVs reg=easy count=0 len=11415841 perMb=0.000\n",
      "[RESULT] hap=paternal mut=all reg=all count=609 len=27297315 perMb=22.310\n",
      "[RESULT] hap=paternal mut=SNVs reg=hard count=8 len=25670320 perMb=0.312\n",
      "[RESULT] hap=paternal mut=Indels reg=all count=840 len=95299862 perMb=8.814\n",
      "[RESULT] hap=paternal mut=SNVs reg=hard count=0 len=1422180 perMb=0.000\n",
      "[RESULT] hap=paternal mut=all reg=hard count=1456 len=24023515 perMb=60.607\n",
      "[RESULT] hap=paternal mut=Indels reg=all count=114 len=12838021 perMb=8.880\n",
      "[RESULT] hap=paternal mut=SNVs reg=all count=20 len=139459947 perMb=0.143\n",
      "[RESULT] hap=maternal mut=Indels reg=easy count=2 len=18817422 perMb=0.106\n",
      "[RESULT] hap=paternal mut=Indels reg=easy count=0 len=11415841 perMb=0.000\n",
      "[RESULT] hap=all mut=Indels reg=easy count=9 len=109064061 perMb=0.083\n",
      "[RESULT] hap=paternal mut=all reg=easy count=0 len=20342029 perMb=0.000\n",
      "[RESULT] hap=paternal mut=Indels reg=hard count=114 len=1422180 perMb=80.159\n",
      "[DONE] summary written to /tmp/mutrate_896r0tio/chr20/mutation_rates_summary.tsv\n",
      "[INFO] cleaning tmpdir: /tmp/mutrate_896r0tio/chr20\n",
      "[RESULT] hap=maternal mut=Indels reg=hard count=343 len=6795117 perMb=50.477\n",
      "[OK] chr20\n",
      "[RESULT] hap=paternal mut=all reg=all count=339 len=26258967 perMb=12.910\n",
      "[RESULT] hap=paternal mut=Indels reg=easy count=4 len=69629542 perMb=0.057\n",
      "[RESULT] hap=paternal mut=all reg=hard count=609 len=6955286 perMb=87.559\n",
      "[RESULT] hap=paternal mut=SNVs reg=all count=7 len=27297315 perMb=0.256\n",
      "[RESULT] hap=paternal mut=all reg=easy count=2 len=13285715 perMb=0.151\n",
      "[RESULT] hap=all mut=Indels reg=hard count=727 len=32674475 perMb=22.250\n",
      "[RESULT] hap=maternal mut=all reg=all count=645 len=128906940 perMb=5.004\n",
      "[RESULT] hap=paternal mut=all reg=hard count=337 len=12973252 perMb=25.977\n",
      "[RESULT] hap=paternal mut=SNVs reg=all count=7 len=26258967 perMb=0.267\n",
      "[RESULT] hap=paternal mut=SNVs reg=easy count=0 len=20342029 perMb=0.000\n",
      "[RESULT] hap=paternal mut=Indels reg=hard count=836 len=25670320 perMb=32.567\n",
      "[DONE] summary written to /tmp/mutrate_896r0tio/chr2/mutation_rates_summary.tsv\n",
      "[INFO] cleaning tmpdir: /tmp/mutrate_896r0tio/chr2\n",
      "[RESULT] hap=paternal mut=SNVs reg=easy count=2 len=115436432 perMb=0.017\n",
      "[RESULT] hap=paternal mut=SNVs reg=easy count=1 len=13285715 perMb=0.075\n",
      "[OK] chr2\n",
      "[RESULT] hap=paternal mut=SNVs reg=hard count=7 len=6955286 perMb=1.006\n",
      "[RESULT] hap=maternal mut=all reg=easy count=21 len=99341133 perMb=0.211\n",
      "[RESULT] hap=paternal mut=SNVs reg=hard count=6 len=12973252 perMb=0.462\n",
      "[RESULT] hap=paternal mut=Indels reg=all count=602 len=27297315 perMb=22.053\n",
      "[RESULT] hap=paternal mut=Indels reg=all count=332 len=26258967 perMb=12.643\n",
      "[RESULT] hap=paternal mut=Indels reg=easy count=1 len=13285715 perMb=0.075\n",
      "[RESULT] hap=paternal mut=Indels reg=easy count=0 len=20342029 perMb=0.000\n",
      "[RESULT] hap=paternal mut=Indels reg=hard count=331 len=12973252 perMb=25.514\n",
      "[DONE] summary written to /tmp/mutrate_896r0tio/chr19/mutation_rates_summary.tsv\n",
      "[INFO] cleaning tmpdir: /tmp/mutrate_896r0tio/chr19\n",
      "[RESULT] hap=maternal mut=all reg=hard count=624 len=29565807 perMb=21.105\n",
      "[RESULT] hap=maternal mut=SNVs reg=all count=36 len=128906940 perMb=0.279\n",
      "[OK] chr19\n",
      "[RESULT] hap=paternal mut=Indels reg=hard count=602 len=6955286 perMb=86.553\n",
      "[DONE] summary written to /tmp/mutrate_896r0tio/chr17/mutation_rates_summary.tsv\n",
      "[INFO] cleaning tmpdir: /tmp/mutrate_896r0tio/chr17\n",
      "[RESULT] hap=paternal mut=SNVs reg=hard count=18 len=24023515 perMb=0.749\n",
      "[RESULT] hap=paternal mut=Indels reg=all count=1443 len=139459947 perMb=10.347\n",
      "[OK] chr17\n",
      "[RESULT] hap=maternal mut=SNVs reg=easy count=14 len=99341133 perMb=0.141\n",
      "[RESULT] hap=maternal mut=SNVs reg=hard count=22 len=29565807 perMb=0.744\n",
      "[RESULT] hap=maternal mut=Indels reg=all count=609 len=128906940 perMb=4.724\n",
      "[RESULT] hap=paternal mut=Indels reg=easy count=7 len=115436432 perMb=0.061\n",
      "[RESULT] hap=maternal mut=Indels reg=easy count=7 len=99341133 perMb=0.070\n",
      "[RESULT] hap=maternal mut=Indels reg=hard count=602 len=29565807 perMb=20.361\n",
      "[RESULT] hap=paternal mut=all reg=all count=129 len=12831596 perMb=10.053\n",
      "[RESULT] hap=paternal mut=all reg=easy count=2 len=9722928 perMb=0.206\n",
      "[RESULT] hap=paternal mut=all reg=hard count=127 len=3108668 perMb=40.854\n",
      "[RESULT] hap=paternal mut=SNVs reg=all count=2 len=12831596 perMb=0.156\n",
      "[RESULT] hap=paternal mut=Indels reg=hard count=1438 len=24023515 perMb=59.858\n",
      "[DONE] summary written to /tmp/mutrate_896r0tio/chr1/mutation_rates_summary.tsv\n",
      "[INFO] cleaning tmpdir: /tmp/mutrate_896r0tio/chr1\n",
      "[RESULT] hap=paternal mut=SNVs reg=easy count=0 len=9722928 perMb=0.000\n",
      "[RESULT] hap=paternal mut=SNVs reg=hard count=2 len=3108668 perMb=0.643\n",
      "[OK] chr1\n",
      "[RESULT] hap=paternal mut=Indels reg=all count=127 len=12831596 perMb=9.897\n",
      "[RESULT] hap=paternal mut=Indels reg=easy count=2 len=9722928 perMb=0.206\n",
      "[RESULT] hap=paternal mut=Indels reg=hard count=125 len=3108668 perMb=40.210\n",
      "[DONE] summary written to /tmp/mutrate_896r0tio/chrX/mutation_rates_summary.tsv\n",
      "[INFO] cleaning tmpdir: /tmp/mutrate_896r0tio/chrX\n",
      "[OK] chrX\n",
      "[INFO] Final concatenated shape: (621, 9)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>haplotype_scope</th>\n",
       "      <th>region_scope</th>\n",
       "      <th>mutation_type</th>\n",
       "      <th>region_bed</th>\n",
       "      <th>length_bp</th>\n",
       "      <th>variant_count</th>\n",
       "      <th>variants_per_Mb</th>\n",
       "      <th>region_scope_len</th>\n",
       "      <th>chromosome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>all</td>\n",
       "      <td>all</td>\n",
       "      <td>all</td>\n",
       "      <td>/tmp/mutrate_896r0tio/chr16/all_regions.bed</td>\n",
       "      <td>75751317</td>\n",
       "      <td>812</td>\n",
       "      <td>10.719286</td>\n",
       "      <td>0</td>\n",
       "      <td>chr16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>all</td>\n",
       "      <td>easy</td>\n",
       "      <td>all</td>\n",
       "      <td>/tmp/mutrate_896r0tio/chr16/both_easy.bed</td>\n",
       "      <td>48527966</td>\n",
       "      <td>3</td>\n",
       "      <td>0.061820</td>\n",
       "      <td>0</td>\n",
       "      <td>chr16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>all</td>\n",
       "      <td>hard</td>\n",
       "      <td>all</td>\n",
       "      <td>/tmp/mutrate_896r0tio/chr16/both_hard.bed</td>\n",
       "      <td>27223351</td>\n",
       "      <td>809</td>\n",
       "      <td>29.717135</td>\n",
       "      <td>0</td>\n",
       "      <td>chr16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>all</td>\n",
       "      <td>all</td>\n",
       "      <td>SNVs</td>\n",
       "      <td>/tmp/mutrate_896r0tio/chr16/all_regions.bed</td>\n",
       "      <td>75751317</td>\n",
       "      <td>19</td>\n",
       "      <td>0.250821</td>\n",
       "      <td>0</td>\n",
       "      <td>chr16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>all</td>\n",
       "      <td>easy</td>\n",
       "      <td>SNVs</td>\n",
       "      <td>/tmp/mutrate_896r0tio/chr16/both_easy.bed</td>\n",
       "      <td>48527966</td>\n",
       "      <td>1</td>\n",
       "      <td>0.020607</td>\n",
       "      <td>0</td>\n",
       "      <td>chr16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>616</th>\n",
       "      <td>paternal</td>\n",
       "      <td>easy</td>\n",
       "      <td>SNVs</td>\n",
       "      <td>/tmp/mutrate_896r0tio/chrX/pat_easy.selected.bed</td>\n",
       "      <td>9722928</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>12831596</td>\n",
       "      <td>chrX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>617</th>\n",
       "      <td>paternal</td>\n",
       "      <td>hard</td>\n",
       "      <td>SNVs</td>\n",
       "      <td>/tmp/mutrate_896r0tio/chrX/pat_hard.selected.bed</td>\n",
       "      <td>3108668</td>\n",
       "      <td>2</td>\n",
       "      <td>0.643362</td>\n",
       "      <td>12831596</td>\n",
       "      <td>chrX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>618</th>\n",
       "      <td>paternal</td>\n",
       "      <td>all</td>\n",
       "      <td>Indels</td>\n",
       "      <td>/tmp/mutrate_896r0tio/chrX/pat_union.bed</td>\n",
       "      <td>12831596</td>\n",
       "      <td>127</td>\n",
       "      <td>9.897444</td>\n",
       "      <td>12831596</td>\n",
       "      <td>chrX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>619</th>\n",
       "      <td>paternal</td>\n",
       "      <td>easy</td>\n",
       "      <td>Indels</td>\n",
       "      <td>/tmp/mutrate_896r0tio/chrX/pat_easy.selected.bed</td>\n",
       "      <td>9722928</td>\n",
       "      <td>2</td>\n",
       "      <td>0.205699</td>\n",
       "      <td>12831596</td>\n",
       "      <td>chrX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>620</th>\n",
       "      <td>paternal</td>\n",
       "      <td>hard</td>\n",
       "      <td>Indels</td>\n",
       "      <td>/tmp/mutrate_896r0tio/chrX/pat_hard.selected.bed</td>\n",
       "      <td>3108668</td>\n",
       "      <td>125</td>\n",
       "      <td>40.210148</td>\n",
       "      <td>12831596</td>\n",
       "      <td>chrX</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>621 rows  9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    haplotype_scope region_scope mutation_type  \\\n",
       "0               all          all           all   \n",
       "1               all         easy           all   \n",
       "2               all         hard           all   \n",
       "3               all          all          SNVs   \n",
       "4               all         easy          SNVs   \n",
       "..              ...          ...           ...   \n",
       "616        paternal         easy          SNVs   \n",
       "617        paternal         hard          SNVs   \n",
       "618        paternal          all        Indels   \n",
       "619        paternal         easy        Indels   \n",
       "620        paternal         hard        Indels   \n",
       "\n",
       "                                           region_bed  length_bp  \\\n",
       "0         /tmp/mutrate_896r0tio/chr16/all_regions.bed   75751317   \n",
       "1           /tmp/mutrate_896r0tio/chr16/both_easy.bed   48527966   \n",
       "2           /tmp/mutrate_896r0tio/chr16/both_hard.bed   27223351   \n",
       "3         /tmp/mutrate_896r0tio/chr16/all_regions.bed   75751317   \n",
       "4           /tmp/mutrate_896r0tio/chr16/both_easy.bed   48527966   \n",
       "..                                                ...        ...   \n",
       "616  /tmp/mutrate_896r0tio/chrX/pat_easy.selected.bed    9722928   \n",
       "617  /tmp/mutrate_896r0tio/chrX/pat_hard.selected.bed    3108668   \n",
       "618          /tmp/mutrate_896r0tio/chrX/pat_union.bed   12831596   \n",
       "619  /tmp/mutrate_896r0tio/chrX/pat_easy.selected.bed    9722928   \n",
       "620  /tmp/mutrate_896r0tio/chrX/pat_hard.selected.bed    3108668   \n",
       "\n",
       "     variant_count  variants_per_Mb  region_scope_len chromosome  \n",
       "0              812        10.719286                 0      chr16  \n",
       "1                3         0.061820                 0      chr16  \n",
       "2              809        29.717135                 0      chr16  \n",
       "3               19         0.250821                 0      chr16  \n",
       "4                1         0.020607                 0      chr16  \n",
       "..             ...              ...               ...        ...  \n",
       "616              0         0.000000          12831596       chrX  \n",
       "617              2         0.643362          12831596       chrX  \n",
       "618            127         9.897444          12831596       chrX  \n",
       "619              2         0.205699          12831596       chrX  \n",
       "620            125        40.210148          12831596       chrX  \n",
       "\n",
       "[621 rows x 9 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "base_tmp: str = os.environ.get(\"SCRATCH\", None) or os.environ.get(\"TMPDIR\", None) or \"/tmp\"\n",
    "tmpdir: Path = Path(tempfile.mkdtemp(prefix=\"mutrate_\", dir=base_tmp))\n",
    "print(\"[INFO] tmpdir:\", tmpdir)\n",
    "\n",
    "results: list[dict] = []\n",
    "with ProcessPoolExecutor(max_workers=16) as ex:\n",
    "    futures = {ex.submit(run_single_chromosome, tmpdir, chrom): chrom for chrom in chromosomes}\n",
    "    for fut in as_completed(futures):\n",
    "        chrom = futures[fut]\n",
    "        try:\n",
    "            df = fut.result()\n",
    "            results.append(df)\n",
    "            print(f\"[OK] {chrom}\")\n",
    "        except Exception as e:\n",
    "            print(f\"[FAIL] {chrom}: {e}\")\n",
    "\n",
    "# concat\n",
    "final_df = pd.concat(results, ignore_index=True)\n",
    "final_df.to_csv(\"mutation_rate_by_chromosome.csv\", index=False)\n",
    "print(\"[INFO] Final concatenated shape:\", final_df.shape)\n",
    "\n",
    "final_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:bioinfo]",
   "language": "python",
   "name": "conda-env-bioinfo-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
